// @typhonjs-fvtt/runtime/_dist/svelte/component/core/index.js
import { SvelteComponent, init, safe_not_equal, flush, append_styles, empty, insert, group_outros, transition_out, check_outros, transition_in, detach, create_component, mount_component, noop, destroy_component, create_slot, element, attr, listen, update_slot_base, get_all_dirty_from_scope, get_slot_changes, add_render_callback, create_in_transition, create_out_transition, HtmlTag, text, append, stop_propagation, prevent_default, action_destroyer, set_data, is_function, run_all, space, destroy_each, component_subscribe, update_keyed_each, destroy_block, binding_callbacks, toggle_class, get_spread_update, get_spread_object, assign, add_resize_listener, bind, add_flush_callback } from 'svelte/internal';
import { getContext, setContext } from 'svelte';
import { s_DEFAULT_TRANSITION, s_DEFAULT_TRANSITION_OPTIONS } from '@typhonjs-fvtt/runtime/svelte/transition';
import { writable } from 'svelte/store';
import { applyStyles, draggable } from '@typhonjs-fvtt/runtime/svelte/action';
import { localize } from '@typhonjs-fvtt/runtime/svelte/helper';
import { isObject, isSvelteComponent, parseSvelteConfig } from '@typhonjs-fvtt/runtime/svelte/util';
import { fade } from 'svelte/transition';

/* src\component\core\TJSContainer.svelte generated by Svelte v3.46.0 */

function add_css$4(target) {
	append_styles(target, "svelte-1s361pr", "p.svelte-1s361pr{color:red;font-size:18px}");
}

function get_each_context$2(ctx, list, i) {
	const child_ctx = ctx.slice();
	child_ctx[2] = list[i];
	return child_ctx;
}

// (12:15) 
function create_if_block_1$3(ctx) {
	let p;

	return {
		c() {
			p = element("p");
			p.textContent = "Container warning: No children.";
			attr(p, "class", "svelte-1s361pr");
		},
		m(target, anchor) {
			insert(target, p, anchor);
		},
		p: noop,
		i: noop,
		o: noop,
		d(detaching) {
			if (detaching) detach(p);
		}
	};
}

// (8:0) {#if Array.isArray(children)}
function create_if_block$5(ctx) {
	let each_1_anchor;
	let current;
	let each_value = /*children*/ ctx[1];
	let each_blocks = [];

	for (let i = 0; i < each_value.length; i += 1) {
		each_blocks[i] = create_each_block$2(get_each_context$2(ctx, each_value, i));
	}

	const out = i => transition_out(each_blocks[i], 1, 1, () => {
		each_blocks[i] = null;
	});

	return {
		c() {
			for (let i = 0; i < each_blocks.length; i += 1) {
				each_blocks[i].c();
			}

			each_1_anchor = empty();
		},
		m(target, anchor) {
			for (let i = 0; i < each_blocks.length; i += 1) {
				each_blocks[i].m(target, anchor);
			}

			insert(target, each_1_anchor, anchor);
			current = true;
		},
		p(ctx, dirty) {
			if (dirty & /*children*/ 2) {
				each_value = /*children*/ ctx[1];
				let i;

				for (i = 0; i < each_value.length; i += 1) {
					const child_ctx = get_each_context$2(ctx, each_value, i);

					if (each_blocks[i]) {
						each_blocks[i].p(child_ctx, dirty);
						transition_in(each_blocks[i], 1);
					} else {
						each_blocks[i] = create_each_block$2(child_ctx);
						each_blocks[i].c();
						transition_in(each_blocks[i], 1);
						each_blocks[i].m(each_1_anchor.parentNode, each_1_anchor);
					}
				}

				group_outros();

				for (i = each_value.length; i < each_blocks.length; i += 1) {
					out(i);
				}

				check_outros();
			}
		},
		i(local) {
			if (current) return;

			for (let i = 0; i < each_value.length; i += 1) {
				transition_in(each_blocks[i]);
			}

			current = true;
		},
		o(local) {
			each_blocks = each_blocks.filter(Boolean);

			for (let i = 0; i < each_blocks.length; i += 1) {
				transition_out(each_blocks[i]);
			}

			current = false;
		},
		d(detaching) {
			destroy_each(each_blocks, detaching);
			if (detaching) detach(each_1_anchor);
		}
	};
}

// (9:4) {#each children as child}
function create_each_block$2(ctx) {
	let switch_instance;
	let switch_instance_anchor;
	let current;
	const switch_instance_spread_levels = [/*child*/ ctx[2].props];
	var switch_value = /*child*/ ctx[2].class;

	function switch_props(ctx) {
		let switch_instance_props = {};

		for (let i = 0; i < switch_instance_spread_levels.length; i += 1) {
			switch_instance_props = assign(switch_instance_props, switch_instance_spread_levels[i]);
		}

		return { props: switch_instance_props };
	}

	if (switch_value) {
		switch_instance = new switch_value(switch_props());
	}

	return {
		c() {
			if (switch_instance) create_component(switch_instance.$$.fragment);
			switch_instance_anchor = empty();
		},
		m(target, anchor) {
			if (switch_instance) {
				mount_component(switch_instance, target, anchor);
			}

			insert(target, switch_instance_anchor, anchor);
			current = true;
		},
		p(ctx, dirty) {
			const switch_instance_changes = (dirty & /*children*/ 2)
			? get_spread_update(switch_instance_spread_levels, [get_spread_object(/*child*/ ctx[2].props)])
			: {};

			if (switch_value !== (switch_value = /*child*/ ctx[2].class)) {
				if (switch_instance) {
					group_outros();
					const old_component = switch_instance;

					transition_out(old_component.$$.fragment, 1, 0, () => {
						destroy_component(old_component, 1);
					});

					check_outros();
				}

				if (switch_value) {
					switch_instance = new switch_value(switch_props());
					create_component(switch_instance.$$.fragment);
					transition_in(switch_instance.$$.fragment, 1);
					mount_component(switch_instance, switch_instance_anchor.parentNode, switch_instance_anchor);
				} else {
					switch_instance = null;
				}
			} else if (switch_value) {
				switch_instance.$set(switch_instance_changes);
			}
		},
		i(local) {
			if (current) return;
			if (switch_instance) transition_in(switch_instance.$$.fragment, local);
			current = true;
		},
		o(local) {
			if (switch_instance) transition_out(switch_instance.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			if (detaching) detach(switch_instance_anchor);
			if (switch_instance) destroy_component(switch_instance, detaching);
		}
	};
}

function create_fragment$a(ctx) {
	let show_if;
	let current_block_type_index;
	let if_block;
	let if_block_anchor;
	let current;
	const if_block_creators = [create_if_block$5, create_if_block_1$3];
	const if_blocks = [];

	function select_block_type(ctx, dirty) {
		if (dirty & /*children*/ 2) show_if = null;
		if (show_if == null) show_if = !!Array.isArray(/*children*/ ctx[1]);
		if (show_if) return 0;
		if (/*warn*/ ctx[0]) return 1;
		return -1;
	}

	if (~(current_block_type_index = select_block_type(ctx, -1))) {
		if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);
	}

	return {
		c() {
			if (if_block) if_block.c();
			if_block_anchor = empty();
		},
		m(target, anchor) {
			if (~current_block_type_index) {
				if_blocks[current_block_type_index].m(target, anchor);
			}

			insert(target, if_block_anchor, anchor);
			current = true;
		},
		p(ctx, [dirty]) {
			let previous_block_index = current_block_type_index;
			current_block_type_index = select_block_type(ctx, dirty);

			if (current_block_type_index === previous_block_index) {
				if (~current_block_type_index) {
					if_blocks[current_block_type_index].p(ctx, dirty);
				}
			} else {
				if (if_block) {
					group_outros();

					transition_out(if_blocks[previous_block_index], 1, 1, () => {
						if_blocks[previous_block_index] = null;
					});

					check_outros();
				}

				if (~current_block_type_index) {
					if_block = if_blocks[current_block_type_index];

					if (!if_block) {
						if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);
						if_block.c();
					} else {
						if_block.p(ctx, dirty);
					}

					transition_in(if_block, 1);
					if_block.m(if_block_anchor.parentNode, if_block_anchor);
				} else {
					if_block = null;
				}
			}
		},
		i(local) {
			if (current) return;
			transition_in(if_block);
			current = true;
		},
		o(local) {
			transition_out(if_block);
			current = false;
		},
		d(detaching) {
			if (~current_block_type_index) {
				if_blocks[current_block_type_index].d(detaching);
			}

			if (detaching) detach(if_block_anchor);
		}
	};
}

function instance$a($$self, $$props, $$invalidate) {
	let { warn = false } = $$props;
	let { children = void 0 } = $$props;

	$$self.$$set = $$props => {
		if ('warn' in $$props) $$invalidate(0, warn = $$props.warn);
		if ('children' in $$props) $$invalidate(1, children = $$props.children);
	};

	return [warn, children];
}

class TJSContainer extends SvelteComponent {
	constructor(options) {
		super();
		init(this, options, instance$a, create_fragment$a, safe_not_equal, { warn: 0, children: 1 }, add_css$4);
	}

	get warn() {
		return this.$$.ctx[0];
	}

	set warn(warn) {
		this.$$set({ warn });
		flush();
	}

	get children() {
		return this.$$.ctx[1];
	}

	set children(children) {
		this.$$set({ children });
		flush();
	}
}

/* src\component\core\TJSComponentShell.svelte generated by Svelte v3.46.0 */

function create_fragment$9(ctx) {
	let tjscontainer;
	let current;

	tjscontainer = new TJSContainer({
			props: {
				children: /*allChildren*/ ctx[0],
				warn: true
			}
		});

	return {
		c() {
			create_component(tjscontainer.$$.fragment);
		},
		m(target, anchor) {
			mount_component(tjscontainer, target, anchor);
			current = true;
		},
		p: noop,
		i(local) {
			if (current) return;
			transition_in(tjscontainer.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(tjscontainer.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(tjscontainer, detaching);
		}
	};
}

function instance$9($$self, $$props, $$invalidate) {
	let { children = void 0 } = $$props;
	const context = getContext('external');

	// This component can host multiple children defined via props or in the TyphonJS SvelteData configuration object
	// that are potentially mounted in the content area. If no children defined then this component mounts any slotted
	// child.
	const allChildren = Array.isArray(children)
	? children
	: typeof context === 'object' ? context.children : void 0;

	$$self.$$set = $$props => {
		if ('children' in $$props) $$invalidate(1, children = $$props.children);
	};

	return [allChildren, children];
}

class TJSComponentShell extends SvelteComponent {
	constructor(options) {
		super();
		init(this, options, instance$9, create_fragment$9, safe_not_equal, { children: 1 });
	}

	get children() {
		return this.$$.ctx[1];
	}

	set children(children) {
		this.$$set({ children });
		flush();
	}
}

/* src\component\core\TJSGlassPane.svelte generated by Svelte v3.46.0 */

function add_css$3(target) {
	append_styles(target, "svelte-71db55", ".tjs-glass-pane.svelte-71db55{position:absolute;overflow:inherit}");
}

function create_fragment$8(ctx) {
	let div;
	let div_intro;
	let div_outro;
	let current;
	let mounted;
	let dispose;
	const default_slot_template = /*#slots*/ ctx[17].default;
	const default_slot = create_slot(default_slot_template, ctx, /*$$scope*/ ctx[16], null);

	return {
		c() {
			div = element("div");
			if (default_slot) default_slot.c();
			attr(div, "id", /*id*/ ctx[4]);
			attr(div, "tabindex", "0");
			attr(div, "class", "tjs-glass-pane svelte-71db55");
		},
		m(target, anchor) {
			insert(target, div, anchor);

			if (default_slot) {
				default_slot.m(div, null);
			}

			/*div_binding*/ ctx[18](div);
			current = true;

			if (!mounted) {
				dispose = listen(div, "keydown", /*swallow*/ ctx[6]);
				mounted = true;
			}
		},
		p(new_ctx, [dirty]) {
			ctx = new_ctx;

			if (default_slot) {
				if (default_slot.p && (!current || dirty & /*$$scope*/ 65536)) {
					update_slot_base(
						default_slot,
						default_slot_template,
						ctx,
						/*$$scope*/ ctx[16],
						!current
						? get_all_dirty_from_scope(/*$$scope*/ ctx[16])
						: get_slot_changes(default_slot_template, /*$$scope*/ ctx[16], dirty, null),
						null
					);
				}
			}

			if (!current || dirty & /*id*/ 16) {
				attr(div, "id", /*id*/ ctx[4]);
			}
		},
		i(local) {
			if (current) return;
			transition_in(default_slot, local);

			add_render_callback(() => {
				if (div_outro) div_outro.end(1);
				div_intro = create_in_transition(div, /*inTransition*/ ctx[0], /*inTransitionOptions*/ ctx[2]);
				div_intro.start();
			});

			current = true;
		},
		o(local) {
			transition_out(default_slot, local);
			if (div_intro) div_intro.invalidate();
			div_outro = create_out_transition(div, /*outTransition*/ ctx[1], /*outTransitionOptions*/ ctx[3]);
			current = false;
		},
		d(detaching) {
			if (detaching) detach(div);
			if (default_slot) default_slot.d(detaching);
			/*div_binding*/ ctx[18](null);
			if (detaching && div_outro) div_outro.end();
			mounted = false;
			dispose();
		}
	};
}

function instance$8($$self, $$props, $$invalidate) {
	let { $$slots: slots = {}, $$scope } = $$props;
	let { id = void 0 } = $$props;
	let { zIndex = Number.MAX_SAFE_INTEGER } = $$props;
	let { background = '#50505080' } = $$props;
	let { captureInput = true } = $$props;
	let { preventDefault = true } = $$props;
	let { stopPropagation = true } = $$props;
	let glassPane;
	let { transition = void 0 } = $$props;
	let { inTransition = s_DEFAULT_TRANSITION } = $$props;
	let { outTransition = s_DEFAULT_TRANSITION } = $$props;
	let { transitionOptions = void 0 } = $$props;
	let { inTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS } = $$props;
	let { outTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS } = $$props;

	// Tracks last transition state.
	let oldTransition = void 0;

	let oldTransitionOptions = void 0;

	// ---------------------------------------------------------------------------------------------------------------
	function swallow(event) {
		if (captureInput) {
			if (preventDefault) {
				event.preventDefault();
			}

			if (stopPropagation) {
				event.stopPropagation();
			}
		}
	}

	function div_binding($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			glassPane = $$value;
			((($$invalidate(5, glassPane), $$invalidate(9, captureInput)), $$invalidate(8, background)), $$invalidate(7, zIndex));
		});
	}

	$$self.$$set = $$props => {
		if ('id' in $$props) $$invalidate(4, id = $$props.id);
		if ('zIndex' in $$props) $$invalidate(7, zIndex = $$props.zIndex);
		if ('background' in $$props) $$invalidate(8, background = $$props.background);
		if ('captureInput' in $$props) $$invalidate(9, captureInput = $$props.captureInput);
		if ('preventDefault' in $$props) $$invalidate(10, preventDefault = $$props.preventDefault);
		if ('stopPropagation' in $$props) $$invalidate(11, stopPropagation = $$props.stopPropagation);
		if ('transition' in $$props) $$invalidate(12, transition = $$props.transition);
		if ('inTransition' in $$props) $$invalidate(0, inTransition = $$props.inTransition);
		if ('outTransition' in $$props) $$invalidate(1, outTransition = $$props.outTransition);
		if ('transitionOptions' in $$props) $$invalidate(13, transitionOptions = $$props.transitionOptions);
		if ('inTransitionOptions' in $$props) $$invalidate(2, inTransitionOptions = $$props.inTransitionOptions);
		if ('outTransitionOptions' in $$props) $$invalidate(3, outTransitionOptions = $$props.outTransitionOptions);
		if ('$$scope' in $$props) $$invalidate(16, $$scope = $$props.$$scope);
	};

	$$self.$$.update = () => {
		if ($$self.$$.dirty & /*glassPane*/ 32) {
			if (glassPane) {
				$$invalidate(5, glassPane.style.maxWidth = '100%', glassPane);
				$$invalidate(5, glassPane.style.maxHeight = '100%', glassPane);
				$$invalidate(5, glassPane.style.width = '100%', glassPane);
				$$invalidate(5, glassPane.style.height = '100%', glassPane);
			}
		}

		if ($$self.$$.dirty & /*glassPane, captureInput*/ 544) {
			if (glassPane) {
				if (captureInput) {
					glassPane.focus();
				}

				$$invalidate(5, glassPane.style.pointerEvents = captureInput ? 'auto' : 'none', glassPane);
			}
		}

		if ($$self.$$.dirty & /*glassPane, background*/ 288) {
			if (glassPane) {
				$$invalidate(5, glassPane.style.background = background, glassPane);
			}
		}

		if ($$self.$$.dirty & /*glassPane, zIndex*/ 160) {
			if (glassPane) {
				$$invalidate(5, glassPane.style.zIndex = zIndex, glassPane);
			}
		}

		if ($$self.$$.dirty & /*oldTransition, transition*/ 20480) {
			// Run this reactive block when the last transition state is not equal to the current state.
			if (oldTransition !== transition) {
				// If transition is defined and not the default transition then set it to both in and out transition otherwise
				// set the default transition to both in & out transitions.
				const newTransition = s_DEFAULT_TRANSITION !== transition && typeof transition === 'function'
				? transition
				: s_DEFAULT_TRANSITION;

				$$invalidate(0, inTransition = newTransition);
				$$invalidate(1, outTransition = newTransition);
				$$invalidate(14, oldTransition = newTransition);
			}
		}

		if ($$self.$$.dirty & /*oldTransitionOptions, transitionOptions*/ 40960) {
			// Run this reactive block when the last transition options state is not equal to the current options state.
			if (oldTransitionOptions !== transitionOptions) {
				const newOptions = transitionOptions !== s_DEFAULT_TRANSITION_OPTIONS && typeof transitionOptions === 'object'
				? transitionOptions
				: s_DEFAULT_TRANSITION_OPTIONS;

				$$invalidate(2, inTransitionOptions = newOptions);
				$$invalidate(3, outTransitionOptions = newOptions);
				$$invalidate(15, oldTransitionOptions = newOptions);
			}
		}

		if ($$self.$$.dirty & /*inTransition*/ 1) {
			// Handle cases if inTransition is unset; assign noop default transition function.
			if (typeof inTransition !== 'function') {
				$$invalidate(0, inTransition = s_DEFAULT_TRANSITION);
			}
		}

		if ($$self.$$.dirty & /*outTransition*/ 2) {
			// Handle cases if outTransition is unset; assign noop default transition function.
			if (typeof outTransition !== 'function') {
				$$invalidate(1, outTransition = s_DEFAULT_TRANSITION);
			}
		}

		if ($$self.$$.dirty & /*inTransitionOptions*/ 4) {
			// Handle cases if inTransitionOptions is unset; assign empty default transition options.
			if (typeof inTransitionOptions !== 'object') {
				$$invalidate(2, inTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS);
			}
		}

		if ($$self.$$.dirty & /*outTransitionOptions*/ 8) {
			// Handle cases if outTransitionOptions is unset; assign empty default transition options.
			if (typeof outTransitionOptions !== 'object') {
				$$invalidate(3, outTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS);
			}
		}
	};

	return [
		inTransition,
		outTransition,
		inTransitionOptions,
		outTransitionOptions,
		id,
		glassPane,
		swallow,
		zIndex,
		background,
		captureInput,
		preventDefault,
		stopPropagation,
		transition,
		transitionOptions,
		oldTransition,
		oldTransitionOptions,
		$$scope,
		slots,
		div_binding
	];
}

class TJSGlassPane extends SvelteComponent {
	constructor(options) {
		super();

		init(
			this,
			options,
			instance$8,
			create_fragment$8,
			safe_not_equal,
			{
				id: 4,
				zIndex: 7,
				background: 8,
				captureInput: 9,
				preventDefault: 10,
				stopPropagation: 11,
				transition: 12,
				inTransition: 0,
				outTransition: 1,
				transitionOptions: 13,
				inTransitionOptions: 2,
				outTransitionOptions: 3
			},
			add_css$3
		);
	}

	get id() {
		return this.$$.ctx[4];
	}

	set id(id) {
		this.$$set({ id });
		flush();
	}

	get zIndex() {
		return this.$$.ctx[7];
	}

	set zIndex(zIndex) {
		this.$$set({ zIndex });
		flush();
	}

	get background() {
		return this.$$.ctx[8];
	}

	set background(background) {
		this.$$set({ background });
		flush();
	}

	get captureInput() {
		return this.$$.ctx[9];
	}

	set captureInput(captureInput) {
		this.$$set({ captureInput });
		flush();
	}

	get preventDefault() {
		return this.$$.ctx[10];
	}

	set preventDefault(preventDefault) {
		this.$$set({ preventDefault });
		flush();
	}

	get stopPropagation() {
		return this.$$.ctx[11];
	}

	set stopPropagation(stopPropagation) {
		this.$$set({ stopPropagation });
		flush();
	}

	get transition() {
		return this.$$.ctx[12];
	}

	set transition(transition) {
		this.$$set({ transition });
		flush();
	}

	get inTransition() {
		return this.$$.ctx[0];
	}

	set inTransition(inTransition) {
		this.$$set({ inTransition });
		flush();
	}

	get outTransition() {
		return this.$$.ctx[1];
	}

	set outTransition(outTransition) {
		this.$$set({ outTransition });
		flush();
	}

	get transitionOptions() {
		return this.$$.ctx[13];
	}

	set transitionOptions(transitionOptions) {
		this.$$set({ transitionOptions });
		flush();
	}

	get inTransitionOptions() {
		return this.$$.ctx[2];
	}

	set inTransitionOptions(inTransitionOptions) {
		this.$$set({ inTransitionOptions });
		flush();
	}

	get outTransitionOptions() {
		return this.$$.ctx[3];
	}

	set outTransitionOptions(outTransitionOptions) {
		this.$$set({ outTransitionOptions });
		flush();
	}
}

/* src\component\core\application\TJSHeaderButton.svelte generated by Svelte v3.46.0 */

function create_fragment$7(ctx) {
	let a;
	let html_tag;
	let t;
	let a_class_value;
	let applyStyles_action;
	let mounted;
	let dispose;

	return {
		c() {
			a = element("a");
			html_tag = new HtmlTag();
			t = text(/*label*/ ctx[2]);
			html_tag.a = t;
			attr(a, "class", a_class_value = "header-button " + /*button*/ ctx[0].class);
		},
		m(target, anchor) {
			insert(target, a, anchor);
			html_tag.m(/*icon*/ ctx[1], a);
			append(a, t);

			if (!mounted) {
				dispose = [
					listen(a, "click", stop_propagation(prevent_default(/*onClick*/ ctx[4]))),
					listen(a, "pointerdown", stop_propagation(prevent_default(pointerdown_handler))),
					listen(a, "dblclick", stop_propagation(prevent_default(dblclick_handler))),
					action_destroyer(applyStyles_action = applyStyles.call(null, a, /*styles*/ ctx[3]))
				];

				mounted = true;
			}
		},
		p(ctx, [dirty]) {
			if (dirty & /*icon*/ 2) html_tag.p(/*icon*/ ctx[1]);
			if (dirty & /*label*/ 4) set_data(t, /*label*/ ctx[2]);

			if (dirty & /*button*/ 1 && a_class_value !== (a_class_value = "header-button " + /*button*/ ctx[0].class)) {
				attr(a, "class", a_class_value);
			}

			if (applyStyles_action && is_function(applyStyles_action.update) && dirty & /*styles*/ 8) applyStyles_action.update.call(null, /*styles*/ ctx[3]);
		},
		i: noop,
		o: noop,
		d(detaching) {
			if (detaching) detach(a);
			mounted = false;
			run_all(dispose);
		}
	};
}

const s_REGEX_HTML$1 = /^\s*<.*>$/;
const pointerdown_handler = () => null;
const dblclick_handler = () => null;

function instance$7($$self, $$props, $$invalidate) {
	let { button } = $$props;
	let icon, label, title, styles;

	function onClick() {
		// Accept either callback or onclick as the function / data to invoke.
		const invoke = button.callback ?? button.onclick;

		if (typeof invoke === 'function') {
			invoke.call(button);
			$$invalidate(0, button); // This provides a reactive update if button data changes.
		}
	}

	$$self.$$set = $$props => {
		if ('button' in $$props) $$invalidate(0, button = $$props.button);
	};

	$$self.$$.update = () => {
		if ($$self.$$.dirty & /*button, title*/ 33) {
			if (button) {
				$$invalidate(5, title = typeof button.title === 'string'
				? localize(button.title)
				: '');

				// Handle icon and treat bare strings as the icon class; otherwise assume the icon is fully formed HTML.
				$$invalidate(1, icon = typeof button.icon !== 'string'
				? void 0
				: s_REGEX_HTML$1.test(button.icon)
					? button.icon
					: `<i class="${button.icon}" title="${title}"></i>`);

				$$invalidate(2, label = typeof button.label === 'string'
				? localize(button.label)
				: '');

				$$invalidate(3, styles = typeof button.styles === 'object'
				? button.styles
				: void 0);
			}
		}
	};

	return [button, icon, label, styles, onClick, title];
}

class TJSHeaderButton extends SvelteComponent {
	constructor(options) {
		super();
		init(this, options, instance$7, create_fragment$7, safe_not_equal, { button: 0 });
	}

	get button() {
		return this.$$.ctx[0];
	}

	set button(button) {
		this.$$set({ button });
		flush();
	}
}

/* src\component\core\application\TJSApplicationHeader.svelte generated by Svelte v3.46.0 */

function get_each_context$1(ctx, list, i) {
	const child_ctx = ctx.slice();
	child_ctx[11] = list[i];
	return child_ctx;
}

// (42:4) {#each $storeHeaderButtons as button}
function create_each_block$1(ctx) {
	let tjsheaderbutton;
	let current;
	tjsheaderbutton = new TJSHeaderButton({ props: { button: /*button*/ ctx[11] } });

	return {
		c() {
			create_component(tjsheaderbutton.$$.fragment);
		},
		m(target, anchor) {
			mount_component(tjsheaderbutton, target, anchor);
			current = true;
		},
		p(ctx, dirty) {
			const tjsheaderbutton_changes = {};
			if (dirty & /*$storeHeaderButtons*/ 8) tjsheaderbutton_changes.button = /*button*/ ctx[11];
			tjsheaderbutton.$set(tjsheaderbutton_changes);
		},
		i(local) {
			if (current) return;
			transition_in(tjsheaderbutton.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(tjsheaderbutton.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(tjsheaderbutton, detaching);
		}
	};
}

function create_fragment$6(ctx) {
	let header;
	let h4;
	let t0_value = localize(/*$storeTitle*/ ctx[2]) + "";
	let t0;
	let t1;
	let draggable_action;
	let minimizable_action;
	let current;
	let mounted;
	let dispose;
	let each_value = /*$storeHeaderButtons*/ ctx[3];
	let each_blocks = [];

	for (let i = 0; i < each_value.length; i += 1) {
		each_blocks[i] = create_each_block$1(get_each_context$1(ctx, each_value, i));
	}

	const out = i => transition_out(each_blocks[i], 1, 1, () => {
		each_blocks[i] = null;
	});

	return {
		c() {
			header = element("header");
			h4 = element("h4");
			t0 = text(t0_value);
			t1 = space();

			for (let i = 0; i < each_blocks.length; i += 1) {
				each_blocks[i].c();
			}

			attr(h4, "class", "window-title");
			attr(header, "class", "window-header flexrow");
		},
		m(target, anchor) {
			insert(target, header, anchor);
			append(header, h4);
			append(h4, t0);
			append(header, t1);

			for (let i = 0; i < each_blocks.length; i += 1) {
				each_blocks[i].m(header, null);
			}

			current = true;

			if (!mounted) {
				dispose = [
					action_destroyer(draggable_action = draggable.call(null, header, {
						position: /*application*/ ctx[4].position,
						active: /*$storeDraggable*/ ctx[0],
						storeDragging: /*storeDragging*/ ctx[7]
					})),
					action_destroyer(minimizable_action = /*minimizable*/ ctx[10].call(null, header, /*$storeMinimizable*/ ctx[1]))
				];

				mounted = true;
			}
		},
		p(ctx, [dirty]) {
			if ((!current || dirty & /*$storeTitle*/ 4) && t0_value !== (t0_value = localize(/*$storeTitle*/ ctx[2]) + "")) set_data(t0, t0_value);

			if (dirty & /*$storeHeaderButtons*/ 8) {
				each_value = /*$storeHeaderButtons*/ ctx[3];
				let i;

				for (i = 0; i < each_value.length; i += 1) {
					const child_ctx = get_each_context$1(ctx, each_value, i);

					if (each_blocks[i]) {
						each_blocks[i].p(child_ctx, dirty);
						transition_in(each_blocks[i], 1);
					} else {
						each_blocks[i] = create_each_block$1(child_ctx);
						each_blocks[i].c();
						transition_in(each_blocks[i], 1);
						each_blocks[i].m(header, null);
					}
				}

				group_outros();

				for (i = each_value.length; i < each_blocks.length; i += 1) {
					out(i);
				}

				check_outros();
			}

			if (draggable_action && is_function(draggable_action.update) && dirty & /*$storeDraggable*/ 1) draggable_action.update.call(null, {
				position: /*application*/ ctx[4].position,
				active: /*$storeDraggable*/ ctx[0],
				storeDragging: /*storeDragging*/ ctx[7]
			});

			if (minimizable_action && is_function(minimizable_action.update) && dirty & /*$storeMinimizable*/ 2) minimizable_action.update.call(null, /*$storeMinimizable*/ ctx[1]);
		},
		i(local) {
			if (current) return;

			for (let i = 0; i < each_value.length; i += 1) {
				transition_in(each_blocks[i]);
			}

			current = true;
		},
		o(local) {
			each_blocks = each_blocks.filter(Boolean);

			for (let i = 0; i < each_blocks.length; i += 1) {
				transition_out(each_blocks[i]);
			}

			current = false;
		},
		d(detaching) {
			if (detaching) detach(header);
			destroy_each(each_blocks, detaching);
			mounted = false;
			run_all(dispose);
		}
	};
}

function instance$6($$self, $$props, $$invalidate) {
	let $storeDraggable;
	let $storeMinimizable;
	let $storeTitle;
	let $storeHeaderButtons;
	const application = getContext('external').application;
	const storeTitle = application.reactive.storeAppOptions.title;
	component_subscribe($$self, storeTitle, value => $$invalidate(2, $storeTitle = value));
	const storeDraggable = application.reactive.storeAppOptions.draggable;
	component_subscribe($$self, storeDraggable, value => $$invalidate(0, $storeDraggable = value));
	const storeDragging = application.reactive.storeUIState.dragging;
	const storeHeaderButtons = application.reactive.storeUIState.headerButtons;
	component_subscribe($$self, storeHeaderButtons, value => $$invalidate(3, $storeHeaderButtons = value));
	const storeMinimizable = application.reactive.storeAppOptions.minimizable;
	component_subscribe($$self, storeMinimizable, value => $$invalidate(1, $storeMinimizable = value));

	function minimizable(node, booleanStore) {
		const callback = application._onToggleMinimize.bind(application);

		function activateListeners() {
			node.addEventListener('dblclick', callback);
		}

		function removeListeners() {
			node.removeEventListener('dblclick', callback);
		}

		if (booleanStore) {
			activateListeners();
		}

		return {
			update: booleanStore => // eslint-disable-line no-shadow
			{
				if (booleanStore) {
					activateListeners();
				} else {
					removeListeners();
				}
			},
			destroy: () => removeListeners()
		};
	}

	return [
		$storeDraggable,
		$storeMinimizable,
		$storeTitle,
		$storeHeaderButtons,
		application,
		storeTitle,
		storeDraggable,
		storeDragging,
		storeHeaderButtons,
		storeMinimizable,
		minimizable
	];
}

class TJSApplicationHeader extends SvelteComponent {
	constructor(options) {
		super();
		init(this, options, instance$6, create_fragment$6, safe_not_equal, {});
	}
}

/**
 * Awaits `requestAnimationFrame` calls by the counter specified. This allows asynchronous applications for direct /
 * inline style modification amongst other direct animation techniques.
 *
 * @param {number}   [cntr=1] - A positive integer greater than 0 for amount of requestAnimationFrames to wait.
 *
 * @returns {Promise<number>} Returns current time equivalent to `performance.now()`.
 */
async function nextAnimationFrame(cntr = 1)
{
   if (!Number.isInteger(cntr) || cntr < 1)
   {
      throw new TypeError(`nextAnimationFrame error: 'cntr' must be a positive integer greater than 0.`);
   }

   let currentTime = performance.now();
   for (;--cntr >= 0;)
   {
      currentTime = await new Promise((resolve) => requestAnimationFrame(resolve));
   }

   return currentTime;
}

/* src\component\core\application\ResizableHandle.svelte generated by Svelte v3.46.0 */

function create_fragment$5(ctx) {
	let div;
	let resizable_action;
	let mounted;
	let dispose;

	return {
		c() {
			div = element("div");
			div.innerHTML = `<i class="fas fa-arrows-alt-h"></i>`;
			attr(div, "class", "window-resizable-handle");
		},
		m(target, anchor) {
			insert(target, div, anchor);
			/*div_binding*/ ctx[10](div);

			if (!mounted) {
				dispose = action_destroyer(resizable_action = /*resizable*/ ctx[6].call(null, div, {
					active: /*$storeResizable*/ ctx[1],
					storeResizing: /*storeResizing*/ ctx[5]
				}));

				mounted = true;
			}
		},
		p(ctx, [dirty]) {
			if (resizable_action && is_function(resizable_action.update) && dirty & /*$storeResizable*/ 2) resizable_action.update.call(null, {
				active: /*$storeResizable*/ ctx[1],
				storeResizing: /*storeResizing*/ ctx[5]
			});
		},
		i: noop,
		o: noop,
		d(detaching) {
			if (detaching) detach(div);
			/*div_binding*/ ctx[10](null);
			mounted = false;
			dispose();
		}
	};
}

function instance$5($$self, $$props, $$invalidate) {
	let $storeElementRoot;
	let $storeMinimized;
	let $storeResizable;
	let { isResizable = false } = $$props;
	const application = getContext('external').application;

	// Allows retrieval of the element root at runtime.
	const storeElementRoot = getContext('storeElementRoot');

	component_subscribe($$self, storeElementRoot, value => $$invalidate(8, $storeElementRoot = value));
	const storeResizable = application.reactive.storeAppOptions.resizable;
	component_subscribe($$self, storeResizable, value => $$invalidate(1, $storeResizable = value));
	const storeMinimized = application.reactive.storeUIState.minimized;
	component_subscribe($$self, storeMinimized, value => $$invalidate(9, $storeMinimized = value));
	const storeResizing = application.reactive.storeUIState.resizing;
	let elementResize;

	/**
 * Provides an action to handle resizing the application shell based on the resizable app option.
 *
 * @param {HTMLElement}       node - The node associated with the action.
 *
 * @param {object}            [opts] - Optional parameters.
 *
 * @param {boolean}           [opts.active=true] - A boolean value; attached to a readable store.
 *
 * @param {Writable<boolean>} [opts.storeResizing] - A writable store that tracks "resizing" state.
 *
 * @returns {{update: Function, destroy: Function}} The action lifecycle methods.
 */
	function resizable(node, { active = true, storeResizing = void 0 } = {}) {
		/**
 * Duplicate the app / Positionable starting position to track differences.
 *
 * @type {object}
 */
		let position = null;

		/**
 * Stores the initial X / Y on drag down.
 *
 * @type {object}
 */
		let initialPosition = {};

		/**
 * Stores the current resizing state and gates the move pointer as the resizing store is not
 * set until the first pointer move.
 *
 * @type {boolean}
 */
		let resizing = false;

		/**
 * Remember event handlers associated with this action so they may be later unregistered.
 *
 * @type {Object}
 */
		const handlers = {
			resizeDown: ['pointerdown', e => onResizePointerDown(e), false],
			resizeMove: ['pointermove', e => onResizePointerMove(e), false],
			resizeUp: ['pointerup', e => onResizePointerUp(e), false]
		};

		/**
 * Activates listeners.
 */
		function activateListeners() {
			// Resize handlers
			node.addEventListener(...handlers.resizeDown);

			$$invalidate(7, isResizable = true);
			node.style.display = 'block';
		}

		/**
 * Removes listeners.
 */
		function removeListeners() {
			if (typeof storeResizing?.set === 'function') {
				storeResizing.set(false);
			}

			// Resize handlers
			node.removeEventListener(...handlers.resizeDown);

			node.removeEventListener(...handlers.resizeMove);
			node.removeEventListener(...handlers.resizeUp);
			node.style.display = 'none';
			$$invalidate(7, isResizable = false);
		}

		// On mount if resizable is true then activate listeners otherwise set element display to `none`.
		if (active) {
			activateListeners();
		} else {
			node.style.display = 'none';
		}

		/**
 * Handle the initial pointer down that activates resizing capture.
 */
		function onResizePointerDown(event) {
			event.preventDefault();
			resizing = false;

			// Record initial position
			position = application.position.get();

			if (position.height === 'auto') {
				position.height = $storeElementRoot.clientHeight;
			}

			if (position.width === 'auto') {
				position.width = $storeElementRoot.clientWidth;
			}

			initialPosition = { x: event.clientX, y: event.clientY };

			// Add temporary handlers
			node.addEventListener(...handlers.resizeMove);

			node.addEventListener(...handlers.resizeUp);
			node.setPointerCapture(event.pointerId);
		}

		/**
 * Sets the width / height of the positionable application.
 */
		async function onResizePointerMove(event) {
			event.preventDefault();
			await nextAnimationFrame();

			if (!resizing && typeof storeResizing?.set === 'function') {
				resizing = true;
				storeResizing.set(true);
			}

			application.position.set({
				width: position.width + (event.clientX - initialPosition.x),
				height: position.height + (event.clientY - initialPosition.y)
			});
		}

		/**
 * Conclude the dragging behavior when the pointer is released setting the final position and
 * removing listeners.
 */
		function onResizePointerUp(event) {
			resizing = false;

			if (typeof storeResizing?.set === 'function') {
				storeResizing.set(false);
			}

			event.preventDefault();
			node.removeEventListener(...handlers.resizeMove);
			node.removeEventListener(...handlers.resizeUp);
			application._onResize(event);
		}

		return {
			update: ({ active }) => // eslint-disable-line no-shadow
			{
				if (active) {
					activateListeners();
				} else {
					removeListeners();
				}
			},
			destroy: () => removeListeners()
		};
	}

	function div_binding($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementResize = $$value;
			((($$invalidate(0, elementResize), $$invalidate(7, isResizable)), $$invalidate(9, $storeMinimized)), $$invalidate(8, $storeElementRoot));
		});
	}

	$$self.$$set = $$props => {
		if ('isResizable' in $$props) $$invalidate(7, isResizable = $$props.isResizable);
	};

	$$self.$$.update = () => {
		if ($$self.$$.dirty & /*elementResize, isResizable, $storeMinimized, $storeElementRoot*/ 897) {
			if (elementResize) {
				// Instead of creating a derived store it is easier to use isResizable and the minimized store below.
				$$invalidate(0, elementResize.style.display = isResizable && !$storeMinimized ? 'block' : 'none', elementResize);

				// Add / remove `resizable` class from element root.
				const elementRoot = $storeElementRoot;

				if (elementRoot) {
					elementRoot.classList[isResizable ? 'add' : 'remove']('resizable');
				}
			}
		}
	};

	return [
		elementResize,
		$storeResizable,
		storeElementRoot,
		storeResizable,
		storeMinimized,
		storeResizing,
		resizable,
		isResizable,
		$storeElementRoot,
		$storeMinimized,
		div_binding
	];
}

class ResizableHandle extends SvelteComponent {
	constructor(options) {
		super();
		init(this, options, instance$5, create_fragment$5, safe_not_equal, { isResizable: 7 });
	}
}

/* src\component\core\application\ApplicationShell.svelte generated by Svelte v3.46.0 */

function add_css$2(target) {
	append_styles(target, "svelte-3vt5in", ".window-app.svelte-3vt5in{overflow:inherit}");
}

// (166:0) {:else}
function create_else_block_1$1(ctx) {
	let div;
	let tjsapplicationheader;
	let t0;
	let section;
	let current_block_type_index;
	let if_block;
	let applyStyles_action;
	let t1;
	let resizablehandle;
	let div_id_value;
	let div_class_value;
	let div_data_appid_value;
	let applyStyles_action_1;
	let div_intro;
	let div_outro;
	let current;
	let mounted;
	let dispose;
	tjsapplicationheader = new TJSApplicationHeader({});
	const if_block_creators = [create_if_block_2$2, create_else_block_2$1];
	const if_blocks = [];

	function select_block_type_2(ctx, dirty) {
		if (Array.isArray(/*allChildren*/ ctx[12])) return 0;
		return 1;
	}

	current_block_type_index = select_block_type_2(ctx);
	if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);
	resizablehandle = new ResizableHandle({});

	return {
		c() {
			div = element("div");
			create_component(tjsapplicationheader.$$.fragment);
			t0 = space();
			section = element("section");
			if_block.c();
			t1 = space();
			create_component(resizablehandle.$$.fragment);
			attr(section, "class", "window-content");
			attr(div, "id", div_id_value = /*application*/ ctx[9].id);
			attr(div, "class", div_class_value = "app window-app " + /*application*/ ctx[9].options.classes.join(' ') + " svelte-3vt5in");
			attr(div, "data-appid", div_data_appid_value = /*application*/ ctx[9].appId);
		},
		m(target, anchor) {
			insert(target, div, anchor);
			mount_component(tjsapplicationheader, div, null);
			append(div, t0);
			append(div, section);
			if_blocks[current_block_type_index].m(section, null);
			/*section_binding_1*/ ctx[24](section);
			append(div, t1);
			mount_component(resizablehandle, div, null);
			/*div_binding_1*/ ctx[25](div);
			current = true;

			if (!mounted) {
				dispose = [
					action_destroyer(applyStyles_action = applyStyles.call(null, section, /*stylesContent*/ ctx[8])),
					listen(div, "pointerdown", /*bringToTop*/ ctx[11], true),
					action_destroyer(applyStyles_action_1 = applyStyles.call(null, div, /*stylesApp*/ ctx[7]))
				];

				mounted = true;
			}
		},
		p(new_ctx, dirty) {
			ctx = new_ctx;
			if_block.p(ctx, dirty);
			if (applyStyles_action && is_function(applyStyles_action.update) && dirty & /*stylesContent*/ 256) applyStyles_action.update.call(null, /*stylesContent*/ ctx[8]);

			if (!current || dirty & /*application*/ 512 && div_id_value !== (div_id_value = /*application*/ ctx[9].id)) {
				attr(div, "id", div_id_value);
			}

			if (!current || dirty & /*application*/ 512 && div_class_value !== (div_class_value = "app window-app " + /*application*/ ctx[9].options.classes.join(' ') + " svelte-3vt5in")) {
				attr(div, "class", div_class_value);
			}

			if (!current || dirty & /*application*/ 512 && div_data_appid_value !== (div_data_appid_value = /*application*/ ctx[9].appId)) {
				attr(div, "data-appid", div_data_appid_value);
			}

			if (applyStyles_action_1 && is_function(applyStyles_action_1.update) && dirty & /*stylesApp*/ 128) applyStyles_action_1.update.call(null, /*stylesApp*/ ctx[7]);
		},
		i(local) {
			if (current) return;
			transition_in(tjsapplicationheader.$$.fragment, local);
			transition_in(if_block);
			transition_in(resizablehandle.$$.fragment, local);

			add_render_callback(() => {
				if (div_outro) div_outro.end(1);
				div_intro = create_in_transition(div, /*inTransition*/ ctx[2], /*inTransitionOptions*/ ctx[4]);
				div_intro.start();
			});

			current = true;
		},
		o(local) {
			transition_out(tjsapplicationheader.$$.fragment, local);
			transition_out(if_block);
			transition_out(resizablehandle.$$.fragment, local);
			if (div_intro) div_intro.invalidate();
			div_outro = create_out_transition(div, /*outTransition*/ ctx[3], /*outTransitionOptions*/ ctx[5]);
			current = false;
		},
		d(detaching) {
			if (detaching) detach(div);
			destroy_component(tjsapplicationheader);
			if_blocks[current_block_type_index].d();
			/*section_binding_1*/ ctx[24](null);
			destroy_component(resizablehandle);
			/*div_binding_1*/ ctx[25](null);
			if (detaching && div_outro) div_outro.end();
			mounted = false;
			run_all(dispose);
		}
	};
}

// (143:0) {#if bindHeightChanged}
function create_if_block$4(ctx) {
	let div;
	let tjsapplicationheader;
	let t0;
	let section;
	let current_block_type_index;
	let if_block;
	let section_resize_listener;
	let applyStyles_action;
	let t1;
	let resizablehandle;
	let div_id_value;
	let div_class_value;
	let div_data_appid_value;
	let div_resize_listener;
	let applyStyles_action_1;
	let div_intro;
	let div_outro;
	let current;
	let mounted;
	let dispose;
	tjsapplicationheader = new TJSApplicationHeader({});
	const if_block_creators = [create_if_block_1$2, create_else_block$3];
	const if_blocks = [];

	function select_block_type_1(ctx, dirty) {
		if (Array.isArray(/*allChildren*/ ctx[12])) return 0;
		return 1;
	}

	current_block_type_index = select_block_type_1(ctx);
	if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);
	resizablehandle = new ResizableHandle({});

	return {
		c() {
			div = element("div");
			create_component(tjsapplicationheader.$$.fragment);
			t0 = space();
			section = element("section");
			if_block.c();
			t1 = space();
			create_component(resizablehandle.$$.fragment);
			attr(section, "class", "window-content");
			add_render_callback(() => /*section_elementresize_handler*/ ctx[21].call(section));
			attr(div, "id", div_id_value = /*application*/ ctx[9].id);
			attr(div, "class", div_class_value = "app window-app " + /*application*/ ctx[9].options.classes.join(' ') + " svelte-3vt5in");
			attr(div, "data-appid", div_data_appid_value = /*application*/ ctx[9].appId);
			add_render_callback(() => /*div_elementresize_handler*/ ctx[22].call(div));
		},
		m(target, anchor) {
			insert(target, div, anchor);
			mount_component(tjsapplicationheader, div, null);
			append(div, t0);
			append(div, section);
			if_blocks[current_block_type_index].m(section, null);
			/*section_binding*/ ctx[20](section);
			section_resize_listener = add_resize_listener(section, /*section_elementresize_handler*/ ctx[21].bind(section));
			append(div, t1);
			mount_component(resizablehandle, div, null);
			div_resize_listener = add_resize_listener(div, /*div_elementresize_handler*/ ctx[22].bind(div));
			/*div_binding*/ ctx[23](div);
			current = true;

			if (!mounted) {
				dispose = [
					action_destroyer(applyStyles_action = applyStyles.call(null, section, /*stylesContent*/ ctx[8])),
					listen(div, "pointerdown", /*bringToTop*/ ctx[11], true),
					action_destroyer(applyStyles_action_1 = applyStyles.call(null, div, /*stylesApp*/ ctx[7]))
				];

				mounted = true;
			}
		},
		p(new_ctx, dirty) {
			ctx = new_ctx;
			if_block.p(ctx, dirty);
			if (applyStyles_action && is_function(applyStyles_action.update) && dirty & /*stylesContent*/ 256) applyStyles_action.update.call(null, /*stylesContent*/ ctx[8]);

			if (!current || dirty & /*application*/ 512 && div_id_value !== (div_id_value = /*application*/ ctx[9].id)) {
				attr(div, "id", div_id_value);
			}

			if (!current || dirty & /*application*/ 512 && div_class_value !== (div_class_value = "app window-app " + /*application*/ ctx[9].options.classes.join(' ') + " svelte-3vt5in")) {
				attr(div, "class", div_class_value);
			}

			if (!current || dirty & /*application*/ 512 && div_data_appid_value !== (div_data_appid_value = /*application*/ ctx[9].appId)) {
				attr(div, "data-appid", div_data_appid_value);
			}

			if (applyStyles_action_1 && is_function(applyStyles_action_1.update) && dirty & /*stylesApp*/ 128) applyStyles_action_1.update.call(null, /*stylesApp*/ ctx[7]);
		},
		i(local) {
			if (current) return;
			transition_in(tjsapplicationheader.$$.fragment, local);
			transition_in(if_block);
			transition_in(resizablehandle.$$.fragment, local);

			add_render_callback(() => {
				if (div_outro) div_outro.end(1);
				div_intro = create_in_transition(div, /*inTransition*/ ctx[2], /*inTransitionOptions*/ ctx[4]);
				div_intro.start();
			});

			current = true;
		},
		o(local) {
			transition_out(tjsapplicationheader.$$.fragment, local);
			transition_out(if_block);
			transition_out(resizablehandle.$$.fragment, local);
			if (div_intro) div_intro.invalidate();
			div_outro = create_out_transition(div, /*outTransition*/ ctx[3], /*outTransitionOptions*/ ctx[5]);
			current = false;
		},
		d(detaching) {
			if (detaching) detach(div);
			destroy_component(tjsapplicationheader);
			if_blocks[current_block_type_index].d();
			/*section_binding*/ ctx[20](null);
			section_resize_listener();
			destroy_component(resizablehandle);
			div_resize_listener();
			/*div_binding*/ ctx[23](null);
			if (detaching && div_outro) div_outro.end();
			mounted = false;
			run_all(dispose);
		}
	};
}

// (179:9) {:else}
function create_else_block_2$1(ctx) {
	let current;
	const default_slot_template = /*#slots*/ ctx[19].default;
	const default_slot = create_slot(default_slot_template, ctx, /*$$scope*/ ctx[18], null);

	return {
		c() {
			if (default_slot) default_slot.c();
		},
		m(target, anchor) {
			if (default_slot) {
				default_slot.m(target, anchor);
			}

			current = true;
		},
		p(ctx, dirty) {
			if (default_slot) {
				if (default_slot.p && (!current || dirty & /*$$scope*/ 262144)) {
					update_slot_base(
						default_slot,
						default_slot_template,
						ctx,
						/*$$scope*/ ctx[18],
						!current
						? get_all_dirty_from_scope(/*$$scope*/ ctx[18])
						: get_slot_changes(default_slot_template, /*$$scope*/ ctx[18], dirty, null),
						null
					);
				}
			}
		},
		i(local) {
			if (current) return;
			transition_in(default_slot, local);
			current = true;
		},
		o(local) {
			transition_out(default_slot, local);
			current = false;
		},
		d(detaching) {
			if (default_slot) default_slot.d(detaching);
		}
	};
}

// (177:9) {#if Array.isArray(allChildren)}
function create_if_block_2$2(ctx) {
	let tjscontainer;
	let current;

	tjscontainer = new TJSContainer({
			props: { children: /*allChildren*/ ctx[12] }
		});

	return {
		c() {
			create_component(tjscontainer.$$.fragment);
		},
		m(target, anchor) {
			mount_component(tjscontainer, target, anchor);
			current = true;
		},
		p: noop,
		i(local) {
			if (current) return;
			transition_in(tjscontainer.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(tjscontainer.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(tjscontainer, detaching);
		}
	};
}

// (160:9) {:else}
function create_else_block$3(ctx) {
	let current;
	const default_slot_template = /*#slots*/ ctx[19].default;
	const default_slot = create_slot(default_slot_template, ctx, /*$$scope*/ ctx[18], null);

	return {
		c() {
			if (default_slot) default_slot.c();
		},
		m(target, anchor) {
			if (default_slot) {
				default_slot.m(target, anchor);
			}

			current = true;
		},
		p(ctx, dirty) {
			if (default_slot) {
				if (default_slot.p && (!current || dirty & /*$$scope*/ 262144)) {
					update_slot_base(
						default_slot,
						default_slot_template,
						ctx,
						/*$$scope*/ ctx[18],
						!current
						? get_all_dirty_from_scope(/*$$scope*/ ctx[18])
						: get_slot_changes(default_slot_template, /*$$scope*/ ctx[18], dirty, null),
						null
					);
				}
			}
		},
		i(local) {
			if (current) return;
			transition_in(default_slot, local);
			current = true;
		},
		o(local) {
			transition_out(default_slot, local);
			current = false;
		},
		d(detaching) {
			if (default_slot) default_slot.d(detaching);
		}
	};
}

// (158:9) {#if Array.isArray(allChildren)}
function create_if_block_1$2(ctx) {
	let tjscontainer;
	let current;

	tjscontainer = new TJSContainer({
			props: { children: /*allChildren*/ ctx[12] }
		});

	return {
		c() {
			create_component(tjscontainer.$$.fragment);
		},
		m(target, anchor) {
			mount_component(tjscontainer, target, anchor);
			current = true;
		},
		p: noop,
		i(local) {
			if (current) return;
			transition_in(tjscontainer.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(tjscontainer.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(tjscontainer, detaching);
		}
	};
}

function create_fragment$4(ctx) {
	let current_block_type_index;
	let if_block;
	let if_block_anchor;
	let current;
	const if_block_creators = [create_if_block$4, create_else_block_1$1];
	const if_blocks = [];

	function select_block_type(ctx, dirty) {
		if (/*bindHeightChanged*/ ctx[10]) return 0;
		return 1;
	}

	current_block_type_index = select_block_type(ctx);
	if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);

	return {
		c() {
			if_block.c();
			if_block_anchor = empty();
		},
		m(target, anchor) {
			if_blocks[current_block_type_index].m(target, anchor);
			insert(target, if_block_anchor, anchor);
			current = true;
		},
		p(ctx, [dirty]) {
			if_block.p(ctx, dirty);
		},
		i(local) {
			if (current) return;
			transition_in(if_block);
			current = true;
		},
		o(local) {
			transition_out(if_block);
			current = false;
		},
		d(detaching) {
			if_blocks[current_block_type_index].d(detaching);
			if (detaching) detach(if_block_anchor);
		}
	};
}

function instance$4($$self, $$props, $$invalidate) {
	let { $$slots: slots = {}, $$scope } = $$props;
	let { elementContent } = $$props;
	let { elementRoot } = $$props;
	let { children = void 0 } = $$props;
	let { heightChanged = false } = $$props;
	let { stylesApp } = $$props;
	let { stylesContent } = $$props;

	// Store the initial `heightChanged` state. If it is truthy then `clientHeight` for the content & root elements
	// are bound to `heightChanged` to signal to any parent component of any change to the client & root.
	const bindHeightChanged = !!heightChanged;

	// If the application is a popOut application then when clicked bring to top. Bound to on pointerdown.
	const bringToTop = () => {
		if (typeof application.options.popOut === 'boolean' && application.options.popOut && application !== ui?.activeWindow) {
			application.bringToTop.call(application);
		}
	};

	// Use a writable store to make `elementContent` and `elementRoot` accessible. A store is used in the case when
	// One root component with an `elementRoot` is replaced with another. Due to timing issues and the onDestroy / outro
	// transitions either of these may be set to null. I will investigate more and file a bug against Svelte.
	if (!getContext('storeElementContent')) {
		setContext('storeElementContent', writable(elementContent));
	}

	if (!getContext('storeElementRoot')) {
		setContext('storeElementRoot', writable(elementRoot));
	}

	const context = getContext('external');

	// Store Foundry Application reference.
	const application = context.application;

	// This component can host multiple children defined via props or in the TyphonJS SvelteData configuration object
	// that are potentially mounted in the content area. If no children defined then this component mounts any slotted
	// child.
	const allChildren = Array.isArray(children)
	? children
	: typeof context === 'object' ? context.children : void 0;

	let { transition = void 0 } = $$props;
	let { inTransition = s_DEFAULT_TRANSITION } = $$props;
	let { outTransition = s_DEFAULT_TRANSITION } = $$props;
	let { transitionOptions = void 0 } = $$props;
	let { inTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS } = $$props;
	let { outTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS } = $$props;

	// Tracks last transition state.
	let oldTransition = void 0;

	let oldTransitionOptions = void 0;

	function section_binding($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementContent = $$value;
			$$invalidate(0, elementContent);
		});
	}

	function section_elementresize_handler() {
		heightChanged = this.clientHeight;
		$$invalidate(6, heightChanged);
	}

	function div_elementresize_handler() {
		heightChanged = this.clientHeight;
		$$invalidate(6, heightChanged);
	}

	function div_binding($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementRoot = $$value;
			$$invalidate(1, elementRoot);
		});
	}

	function section_binding_1($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementContent = $$value;
			$$invalidate(0, elementContent);
		});
	}

	function div_binding_1($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementRoot = $$value;
			$$invalidate(1, elementRoot);
		});
	}

	$$self.$$set = $$props => {
		if ('elementContent' in $$props) $$invalidate(0, elementContent = $$props.elementContent);
		if ('elementRoot' in $$props) $$invalidate(1, elementRoot = $$props.elementRoot);
		if ('children' in $$props) $$invalidate(13, children = $$props.children);
		if ('heightChanged' in $$props) $$invalidate(6, heightChanged = $$props.heightChanged);
		if ('stylesApp' in $$props) $$invalidate(7, stylesApp = $$props.stylesApp);
		if ('stylesContent' in $$props) $$invalidate(8, stylesContent = $$props.stylesContent);
		if ('transition' in $$props) $$invalidate(14, transition = $$props.transition);
		if ('inTransition' in $$props) $$invalidate(2, inTransition = $$props.inTransition);
		if ('outTransition' in $$props) $$invalidate(3, outTransition = $$props.outTransition);
		if ('transitionOptions' in $$props) $$invalidate(15, transitionOptions = $$props.transitionOptions);
		if ('inTransitionOptions' in $$props) $$invalidate(4, inTransitionOptions = $$props.inTransitionOptions);
		if ('outTransitionOptions' in $$props) $$invalidate(5, outTransitionOptions = $$props.outTransitionOptions);
		if ('$$scope' in $$props) $$invalidate(18, $$scope = $$props.$$scope);
	};

	$$self.$$.update = () => {
		if ($$self.$$.dirty & /*elementContent*/ 1) {
			// Only update the `elementContent` store if the new `elementContent` is not null or undefined.
			if (elementContent !== void 0 && elementContent !== null) {
				getContext('storeElementContent').set(elementContent);
			}
		}

		if ($$self.$$.dirty & /*elementRoot*/ 2) {
			// Only update the `elementRoot` store if the new `elementRoot` is not null or undefined.
			if (elementRoot !== void 0 && elementRoot !== null) {
				getContext('storeElementRoot').set(elementRoot);
			}
		}

		if ($$self.$$.dirty & /*oldTransition, transition*/ 81920) {
			// Run this reactive block when the last transition state is not equal to the current state.
			if (oldTransition !== transition) {
				// If transition is defined and not the default transition then set it to both in and out transition otherwise
				// set the default transition to both in & out transitions.
				const newTransition = s_DEFAULT_TRANSITION !== transition && typeof transition === 'function'
				? transition
				: s_DEFAULT_TRANSITION;

				$$invalidate(2, inTransition = newTransition);
				$$invalidate(3, outTransition = newTransition);
				$$invalidate(16, oldTransition = newTransition);
			}
		}

		if ($$self.$$.dirty & /*oldTransitionOptions, transitionOptions*/ 163840) {
			// Run this reactive block when the last transition options state is not equal to the current options state.
			if (oldTransitionOptions !== transitionOptions) {
				const newOptions = transitionOptions !== s_DEFAULT_TRANSITION_OPTIONS && typeof transitionOptions === 'object'
				? transitionOptions
				: s_DEFAULT_TRANSITION_OPTIONS;

				$$invalidate(4, inTransitionOptions = newOptions);
				$$invalidate(5, outTransitionOptions = newOptions);
				$$invalidate(17, oldTransitionOptions = newOptions);
			}
		}

		if ($$self.$$.dirty & /*inTransition*/ 4) {
			// Handle cases if inTransition is unset; assign noop default transition function.
			if (typeof inTransition !== 'function') {
				$$invalidate(2, inTransition = s_DEFAULT_TRANSITION);
			}
		}

		if ($$self.$$.dirty & /*outTransition, application*/ 520) {
			{
				// Handle cases if outTransition is unset; assign noop default transition function.
				if (typeof outTransition !== 'function') {
					$$invalidate(3, outTransition = s_DEFAULT_TRANSITION);
				}

				// Set jquery close animation to either run or not when an out transition is changed.
				if (application && typeof application?.options?.defaultCloseAnimation === 'boolean') {
					$$invalidate(9, application.options.defaultCloseAnimation = outTransition === s_DEFAULT_TRANSITION, application);
				}
			}
		}

		if ($$self.$$.dirty & /*inTransitionOptions*/ 16) {
			// Handle cases if inTransitionOptions is unset; assign empty default transition options.
			if (typeof inTransitionOptions !== 'object') {
				$$invalidate(4, inTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS);
			}
		}

		if ($$self.$$.dirty & /*outTransitionOptions*/ 32) {
			// Handle cases if outTransitionOptions is unset; assign empty default transition options.
			if (typeof outTransitionOptions !== 'object') {
				$$invalidate(5, outTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS);
			}
		}
	};

	return [
		elementContent,
		elementRoot,
		inTransition,
		outTransition,
		inTransitionOptions,
		outTransitionOptions,
		heightChanged,
		stylesApp,
		stylesContent,
		application,
		bindHeightChanged,
		bringToTop,
		allChildren,
		children,
		transition,
		transitionOptions,
		oldTransition,
		oldTransitionOptions,
		$$scope,
		slots,
		section_binding,
		section_elementresize_handler,
		div_elementresize_handler,
		div_binding,
		section_binding_1,
		div_binding_1
	];
}

class ApplicationShell extends SvelteComponent {
	constructor(options) {
		super();

		init(
			this,
			options,
			instance$4,
			create_fragment$4,
			safe_not_equal,
			{
				elementContent: 0,
				elementRoot: 1,
				children: 13,
				heightChanged: 6,
				stylesApp: 7,
				stylesContent: 8,
				transition: 14,
				inTransition: 2,
				outTransition: 3,
				transitionOptions: 15,
				inTransitionOptions: 4,
				outTransitionOptions: 5
			},
			add_css$2
		);
	}

	get elementContent() {
		return this.$$.ctx[0];
	}

	set elementContent(elementContent) {
		this.$$set({ elementContent });
		flush();
	}

	get elementRoot() {
		return this.$$.ctx[1];
	}

	set elementRoot(elementRoot) {
		this.$$set({ elementRoot });
		flush();
	}

	get children() {
		return this.$$.ctx[13];
	}

	set children(children) {
		this.$$set({ children });
		flush();
	}

	get heightChanged() {
		return this.$$.ctx[6];
	}

	set heightChanged(heightChanged) {
		this.$$set({ heightChanged });
		flush();
	}

	get stylesApp() {
		return this.$$.ctx[7];
	}

	set stylesApp(stylesApp) {
		this.$$set({ stylesApp });
		flush();
	}

	get stylesContent() {
		return this.$$.ctx[8];
	}

	set stylesContent(stylesContent) {
		this.$$set({ stylesContent });
		flush();
	}

	get transition() {
		return this.$$.ctx[14];
	}

	set transition(transition) {
		this.$$set({ transition });
		flush();
	}

	get inTransition() {
		return this.$$.ctx[2];
	}

	set inTransition(inTransition) {
		this.$$set({ inTransition });
		flush();
	}

	get outTransition() {
		return this.$$.ctx[3];
	}

	set outTransition(outTransition) {
		this.$$set({ outTransition });
		flush();
	}

	get transitionOptions() {
		return this.$$.ctx[15];
	}

	set transitionOptions(transitionOptions) {
		this.$$set({ transitionOptions });
		flush();
	}

	get inTransitionOptions() {
		return this.$$.ctx[4];
	}

	set inTransitionOptions(inTransitionOptions) {
		this.$$set({ inTransitionOptions });
		flush();
	}

	get outTransitionOptions() {
		return this.$$.ctx[5];
	}

	set outTransitionOptions(outTransitionOptions) {
		this.$$set({ outTransitionOptions });
		flush();
	}
}

/* src\component\core\application\EmptyApplicationShell.svelte generated by Svelte v3.46.0 */

function create_else_block$2(ctx) {
	let div;
	let div_id_value;
	let div_class_value;
	let div_data_appid_value;
	let div_intro;
	let div_outro;
	let current;

	return {
		c() {
			div = element("div");
			attr(div, "id", div_id_value = /*application*/ ctx[6].id);
			attr(div, "class", div_class_value = /*application*/ ctx[6].options.classes.join(' '));
			attr(div, "data-appid", div_data_appid_value = /*application*/ ctx[6].appId);
		},
		m(target, anchor) {
			insert(target, div, anchor);
			/*div_binding_1*/ ctx[15](div);
			current = true;
		},
		p(new_ctx, dirty) {
			ctx = new_ctx;

			if (!current || dirty & /*application*/ 64 && div_id_value !== (div_id_value = /*application*/ ctx[6].id)) {
				attr(div, "id", div_id_value);
			}

			if (!current || dirty & /*application*/ 64 && div_class_value !== (div_class_value = /*application*/ ctx[6].options.classes.join(' '))) {
				attr(div, "class", div_class_value);
			}

			if (!current || dirty & /*application*/ 64 && div_data_appid_value !== (div_data_appid_value = /*application*/ ctx[6].appId)) {
				attr(div, "data-appid", div_data_appid_value);
			}
		},
		i(local) {
			if (current) return;

			add_render_callback(() => {
				if (div_outro) div_outro.end(1);
				div_intro = create_in_transition(div, /*inTransition*/ ctx[1], /*inTransitionOptions*/ ctx[3]);
				div_intro.start();
			});

			current = true;
		},
		o(local) {
			if (div_intro) div_intro.invalidate();
			div_outro = create_out_transition(div, /*outTransition*/ ctx[2], /*outTransitionOptions*/ ctx[4]);
			current = false;
		},
		d(detaching) {
			if (detaching) detach(div);
			/*div_binding_1*/ ctx[15](null);
			if (detaching && div_outro) div_outro.end();
		}
	};
}

// (117:0) {#if bindHeightChanged}
function create_if_block$3(ctx) {
	let div;
	let div_id_value;
	let div_class_value;
	let div_data_appid_value;
	let div_resize_listener;
	let div_intro;
	let div_outro;
	let current;

	return {
		c() {
			div = element("div");
			attr(div, "id", div_id_value = /*application*/ ctx[6].id);
			attr(div, "class", div_class_value = /*application*/ ctx[6].options.classes.join(' '));
			attr(div, "data-appid", div_data_appid_value = /*application*/ ctx[6].appId);
			add_render_callback(() => /*div_elementresize_handler*/ ctx[13].call(div));
		},
		m(target, anchor) {
			insert(target, div, anchor);
			div_resize_listener = add_resize_listener(div, /*div_elementresize_handler*/ ctx[13].bind(div));
			/*div_binding*/ ctx[14](div);
			current = true;
		},
		p(new_ctx, dirty) {
			ctx = new_ctx;

			if (!current || dirty & /*application*/ 64 && div_id_value !== (div_id_value = /*application*/ ctx[6].id)) {
				attr(div, "id", div_id_value);
			}

			if (!current || dirty & /*application*/ 64 && div_class_value !== (div_class_value = /*application*/ ctx[6].options.classes.join(' '))) {
				attr(div, "class", div_class_value);
			}

			if (!current || dirty & /*application*/ 64 && div_data_appid_value !== (div_data_appid_value = /*application*/ ctx[6].appId)) {
				attr(div, "data-appid", div_data_appid_value);
			}
		},
		i(local) {
			if (current) return;

			add_render_callback(() => {
				if (div_outro) div_outro.end(1);
				div_intro = create_in_transition(div, /*inTransition*/ ctx[1], /*inTransitionOptions*/ ctx[3]);
				div_intro.start();
			});

			current = true;
		},
		o(local) {
			if (div_intro) div_intro.invalidate();
			div_outro = create_out_transition(div, /*outTransition*/ ctx[2], /*outTransitionOptions*/ ctx[4]);
			current = false;
		},
		d(detaching) {
			if (detaching) detach(div);
			div_resize_listener();
			/*div_binding*/ ctx[14](null);
			if (detaching && div_outro) div_outro.end();
		}
	};
}

function create_fragment$3(ctx) {
	let current_block_type_index;
	let if_block;
	let if_block_anchor;
	let current;
	const if_block_creators = [create_if_block$3, create_else_block$2];
	const if_blocks = [];

	function select_block_type(ctx, dirty) {
		if (/*bindHeightChanged*/ ctx[7]) return 0;
		return 1;
	}

	current_block_type_index = select_block_type(ctx);
	if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);

	return {
		c() {
			if_block.c();
			if_block_anchor = empty();
		},
		m(target, anchor) {
			if_blocks[current_block_type_index].m(target, anchor);
			insert(target, if_block_anchor, anchor);
			current = true;
		},
		p(ctx, [dirty]) {
			if_block.p(ctx, dirty);
		},
		i(local) {
			if (current) return;
			transition_in(if_block);
			current = true;
		},
		o(local) {
			transition_out(if_block);
			current = false;
		},
		d(detaching) {
			if_blocks[current_block_type_index].d(detaching);
			if (detaching) detach(if_block_anchor);
		}
	};
}

function instance$3($$self, $$props, $$invalidate) {
	let { elementContent } = $$props;
	let { elementRoot } = $$props;
	let { heightChanged = false } = $$props;

	// Store the initial `heightChanged` state. If it is truthy then `clientHeight` for the content & root elements
	// are bound to `heightChanged` to signal to any parent component of any change to the client & root.
	const bindHeightChanged = !!heightChanged;

	// Use a writable store to make `elementContent` and `elementRoot` accessible. A store is used in the case when
	// One root component with an `elementRoot` is replaced with another. Due to timing issues and the onDestroy / outro
	// transitions either of these may be set to null. I will investigate more and file a bug against Svelte.
	if (!getContext('storeElementContent')) {
		setContext('storeElementContent', writable(elementContent));
	}

	if (!getContext('storeElementRoot')) {
		setContext('storeElementRoot', writable(elementRoot));
	}

	const context = getContext('external');

	// Store Foundry Application reference.
	const application = context.application;

	let { transition = void 0 } = $$props;
	let { inTransition = s_DEFAULT_TRANSITION } = $$props;
	let { outTransition = s_DEFAULT_TRANSITION } = $$props;
	let { transitionOptions = void 0 } = $$props;
	let { inTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS } = $$props;
	let { outTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS } = $$props;

	// Tracks last transition state.
	let oldTransition = void 0;

	let oldTransitionOptions = void 0;

	function div_elementresize_handler() {
		heightChanged = this.clientHeight;
		$$invalidate(5, heightChanged);
	}

	function div_binding($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementRoot = $$value;
			$$invalidate(0, elementRoot);
		});
	}

	function div_binding_1($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementRoot = $$value;
			$$invalidate(0, elementRoot);
		});
	}

	$$self.$$set = $$props => {
		if ('elementContent' in $$props) $$invalidate(8, elementContent = $$props.elementContent);
		if ('elementRoot' in $$props) $$invalidate(0, elementRoot = $$props.elementRoot);
		if ('heightChanged' in $$props) $$invalidate(5, heightChanged = $$props.heightChanged);
		if ('transition' in $$props) $$invalidate(9, transition = $$props.transition);
		if ('inTransition' in $$props) $$invalidate(1, inTransition = $$props.inTransition);
		if ('outTransition' in $$props) $$invalidate(2, outTransition = $$props.outTransition);
		if ('transitionOptions' in $$props) $$invalidate(10, transitionOptions = $$props.transitionOptions);
		if ('inTransitionOptions' in $$props) $$invalidate(3, inTransitionOptions = $$props.inTransitionOptions);
		if ('outTransitionOptions' in $$props) $$invalidate(4, outTransitionOptions = $$props.outTransitionOptions);
	};

	$$self.$$.update = () => {
		if ($$self.$$.dirty & /*elementRoot*/ 1) {
			if (elementRoot) {
				$$invalidate(8, elementContent = elementRoot);
			}
		}

		if ($$self.$$.dirty & /*elementContent*/ 256) {
			// Only update the `elementContent` store if the new `elementContent` is not null or undefined.
			if (elementContent !== void 0 && elementContent !== null) {
				getContext('storeElementContent').set(elementContent);
			}
		}

		if ($$self.$$.dirty & /*elementRoot*/ 1) {
			// Only update the `elementRoot` store if the new `elementRoot` is not null or undefined.
			if (elementRoot !== void 0 && elementRoot !== null) {
				getContext('storeElementRoot').set(elementRoot);
			}
		}

		if ($$self.$$.dirty & /*oldTransition, transition*/ 2560) {
			// Run this reactive block when the last transition state is not equal to the current state.
			if (oldTransition !== transition) {
				// If transition is defined and not the default transition then set it to both in and out transition otherwise
				// set the default transition to both in & out transitions.
				const newTransition = s_DEFAULT_TRANSITION !== transition && typeof transition === 'function'
				? transition
				: s_DEFAULT_TRANSITION;

				$$invalidate(1, inTransition = newTransition);
				$$invalidate(2, outTransition = newTransition);
				$$invalidate(11, oldTransition = newTransition);
			}
		}

		if ($$self.$$.dirty & /*oldTransitionOptions, transitionOptions*/ 5120) {
			// Run this reactive block when the last transition options state is not equal to the current options state.
			if (oldTransitionOptions !== transitionOptions) {
				const newOptions = transitionOptions !== s_DEFAULT_TRANSITION_OPTIONS && typeof transitionOptions === 'object'
				? transitionOptions
				: s_DEFAULT_TRANSITION_OPTIONS;

				$$invalidate(3, inTransitionOptions = newOptions);
				$$invalidate(4, outTransitionOptions = newOptions);
				$$invalidate(12, oldTransitionOptions = newOptions);
			}
		}

		if ($$self.$$.dirty & /*inTransition*/ 2) {
			// Handle cases if inTransition is unset; assign noop default transition function.
			if (typeof inTransition !== 'function') {
				$$invalidate(1, inTransition = s_DEFAULT_TRANSITION);
			}
		}

		if ($$self.$$.dirty & /*outTransition, application*/ 68) {
			{
				// Handle cases if outTransition is unset; assign noop default transition function.
				if (typeof outTransition !== 'function') {
					$$invalidate(2, outTransition = s_DEFAULT_TRANSITION);
				}

				// Set jquery close animation to either run or not when an out transition is changed.
				if (application && typeof application?.options?.defaultCloseAnimation === 'boolean') {
					$$invalidate(6, application.options.defaultCloseAnimation = outTransition === s_DEFAULT_TRANSITION, application);
				}
			}
		}

		if ($$self.$$.dirty & /*inTransitionOptions*/ 8) {
			// Handle cases if inTransitionOptions is unset; assign empty default transition options.
			if (typeof inTransitionOptions !== 'object') {
				$$invalidate(3, inTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS);
			}
		}

		if ($$self.$$.dirty & /*outTransitionOptions*/ 16) {
			// Handle cases if outTransitionOptions is unset; assign empty default transition options.
			if (typeof outTransitionOptions !== 'object') {
				$$invalidate(4, outTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS);
			}
		}
	};

	return [
		elementRoot,
		inTransition,
		outTransition,
		inTransitionOptions,
		outTransitionOptions,
		heightChanged,
		application,
		bindHeightChanged,
		elementContent,
		transition,
		transitionOptions,
		oldTransition,
		oldTransitionOptions,
		div_elementresize_handler,
		div_binding,
		div_binding_1
	];
}

class EmptyApplicationShell extends SvelteComponent {
	constructor(options) {
		super();

		init(this, options, instance$3, create_fragment$3, safe_not_equal, {
			elementContent: 8,
			elementRoot: 0,
			heightChanged: 5,
			transition: 9,
			inTransition: 1,
			outTransition: 2,
			transitionOptions: 10,
			inTransitionOptions: 3,
			outTransitionOptions: 4
		});
	}

	get elementContent() {
		return this.$$.ctx[8];
	}

	set elementContent(elementContent) {
		this.$$set({ elementContent });
		flush();
	}

	get elementRoot() {
		return this.$$.ctx[0];
	}

	set elementRoot(elementRoot) {
		this.$$set({ elementRoot });
		flush();
	}

	get heightChanged() {
		return this.$$.ctx[5];
	}

	set heightChanged(heightChanged) {
		this.$$set({ heightChanged });
		flush();
	}

	get transition() {
		return this.$$.ctx[9];
	}

	set transition(transition) {
		this.$$set({ transition });
		flush();
	}

	get inTransition() {
		return this.$$.ctx[1];
	}

	set inTransition(inTransition) {
		this.$$set({ inTransition });
		flush();
	}

	get outTransition() {
		return this.$$.ctx[2];
	}

	set outTransition(outTransition) {
		this.$$set({ outTransition });
		flush();
	}

	get transitionOptions() {
		return this.$$.ctx[10];
	}

	set transitionOptions(transitionOptions) {
		this.$$set({ transitionOptions });
		flush();
	}

	get inTransitionOptions() {
		return this.$$.ctx[3];
	}

	set inTransitionOptions(inTransitionOptions) {
		this.$$set({ inTransitionOptions });
		flush();
	}

	get outTransitionOptions() {
		return this.$$.ctx[4];
	}

	set outTransitionOptions(outTransitionOptions) {
		this.$$set({ outTransitionOptions });
		flush();
	}
}

/* src\component\core\application\TJSApplicationShell.svelte generated by Svelte v3.46.0 */

function add_css$1(target) {
	append_styles(target, "svelte-9xueci", ".tjs-app{max-height:100%;background:url(/ui/denim075.png) repeat;border-radius:5px;box-shadow:0 0 20px #000;margin:3px 0;padding:0.5em;color:#f0f0e0;z-index:95;overflow:inherit}.tjs-window-app{display:flex;flex-direction:column;flex-wrap:nowrap;justify-content:flex-start;position:absolute;box-shadow:0 0 20px #000;padding:0;z-index:95}.tjs-window-app > *{flex:1}.tjs-window-app > .flex0{display:block;flex:0}.tjs-window-app > .flex1{flex:1}.tjs-window-app > .flex2{flex:2}.tjs-window-app > .flex3{flex:3}.tjs-window-app .window-header{flex:0 0 30px;overflow:hidden;padding:0 8px;line-height:30px;border-bottom:1px solid #000}.tjs-window-app .window-header .window-title{margin:0;word-break:break-all}.tjs-window-app .window-header a{flex:none;margin:0 0 0 8px}.tjs-window-app .window-header i[class^=fa]{margin-right:3px}.tjs-window-app.minimized .window-header{border:1px solid #000}.tjs-window-app .window-content{display:flex;flex-direction:column;flex-wrap:nowrap;justify-content:flex-start;padding:8px;color:#191813;overflow-y:auto;overflow-x:hidden}.window-app .window-content > *{flex:1}.window-app .window-content > .flex0{display:block;flex:0}.window-app .window-content > .flex1{flex:1}.window-app .window-content > .flex2{flex:2}.window-app .window-content > .flex3{flex:3}.window-app.zhover{z-index:calc(var(--z-index-window) + 1)}.tjs-window-app .window-resizable-handle{width:20px;height:20px;position:absolute;bottom:-1px;right:0;background:#444;padding:2px;border:1px solid #111;border-radius:4px 0 0 0}.tjs-window-app .window-resizable-handle i.fas{transform:rotate(45deg)}.window-app.minimized .window-resizable-handle{display:none}");
}

// (166:0) {:else}
function create_else_block_1(ctx) {
	let div;
	let tjsapplicationheader;
	let t0;
	let section;
	let current_block_type_index;
	let if_block;
	let applyStyles_action;
	let t1;
	let resizablehandle;
	let div_id_value;
	let div_class_value;
	let div_data_appid_value;
	let applyStyles_action_1;
	let div_intro;
	let div_outro;
	let current;
	let mounted;
	let dispose;
	tjsapplicationheader = new TJSApplicationHeader({});
	const if_block_creators = [create_if_block_2$1, create_else_block_2];
	const if_blocks = [];

	function select_block_type_2(ctx, dirty) {
		if (Array.isArray(/*allChildren*/ ctx[12])) return 0;
		return 1;
	}

	current_block_type_index = select_block_type_2(ctx);
	if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);
	resizablehandle = new ResizableHandle({});

	return {
		c() {
			div = element("div");
			create_component(tjsapplicationheader.$$.fragment);
			t0 = space();
			section = element("section");
			if_block.c();
			t1 = space();
			create_component(resizablehandle.$$.fragment);
			attr(section, "class", "window-content");
			attr(div, "id", div_id_value = /*application*/ ctx[9].id);
			attr(div, "class", div_class_value = "tjs-app tjs-window-app " + /*application*/ ctx[9].options.classes.join(' '));
			attr(div, "data-appid", div_data_appid_value = /*application*/ ctx[9].appId);
		},
		m(target, anchor) {
			insert(target, div, anchor);
			mount_component(tjsapplicationheader, div, null);
			append(div, t0);
			append(div, section);
			if_blocks[current_block_type_index].m(section, null);
			/*section_binding_1*/ ctx[24](section);
			append(div, t1);
			mount_component(resizablehandle, div, null);
			/*div_binding_1*/ ctx[25](div);
			current = true;

			if (!mounted) {
				dispose = [
					action_destroyer(applyStyles_action = applyStyles.call(null, section, /*stylesContent*/ ctx[8])),
					listen(div, "pointerdown", /*bringToTop*/ ctx[11], true),
					action_destroyer(applyStyles_action_1 = applyStyles.call(null, div, /*stylesApp*/ ctx[7]))
				];

				mounted = true;
			}
		},
		p(new_ctx, dirty) {
			ctx = new_ctx;
			if_block.p(ctx, dirty);
			if (applyStyles_action && is_function(applyStyles_action.update) && dirty & /*stylesContent*/ 256) applyStyles_action.update.call(null, /*stylesContent*/ ctx[8]);

			if (!current || dirty & /*application*/ 512 && div_id_value !== (div_id_value = /*application*/ ctx[9].id)) {
				attr(div, "id", div_id_value);
			}

			if (!current || dirty & /*application*/ 512 && div_class_value !== (div_class_value = "tjs-app tjs-window-app " + /*application*/ ctx[9].options.classes.join(' '))) {
				attr(div, "class", div_class_value);
			}

			if (!current || dirty & /*application*/ 512 && div_data_appid_value !== (div_data_appid_value = /*application*/ ctx[9].appId)) {
				attr(div, "data-appid", div_data_appid_value);
			}

			if (applyStyles_action_1 && is_function(applyStyles_action_1.update) && dirty & /*stylesApp*/ 128) applyStyles_action_1.update.call(null, /*stylesApp*/ ctx[7]);
		},
		i(local) {
			if (current) return;
			transition_in(tjsapplicationheader.$$.fragment, local);
			transition_in(if_block);
			transition_in(resizablehandle.$$.fragment, local);

			add_render_callback(() => {
				if (div_outro) div_outro.end(1);
				div_intro = create_in_transition(div, /*inTransition*/ ctx[2], /*inTransitionOptions*/ ctx[4]);
				div_intro.start();
			});

			current = true;
		},
		o(local) {
			transition_out(tjsapplicationheader.$$.fragment, local);
			transition_out(if_block);
			transition_out(resizablehandle.$$.fragment, local);
			if (div_intro) div_intro.invalidate();
			div_outro = create_out_transition(div, /*outTransition*/ ctx[3], /*outTransitionOptions*/ ctx[5]);
			current = false;
		},
		d(detaching) {
			if (detaching) detach(div);
			destroy_component(tjsapplicationheader);
			if_blocks[current_block_type_index].d();
			/*section_binding_1*/ ctx[24](null);
			destroy_component(resizablehandle);
			/*div_binding_1*/ ctx[25](null);
			if (detaching && div_outro) div_outro.end();
			mounted = false;
			run_all(dispose);
		}
	};
}

// (143:0) {#if bindHeightChanged}
function create_if_block$2(ctx) {
	let div;
	let tjsapplicationheader;
	let t0;
	let section;
	let current_block_type_index;
	let if_block;
	let section_resize_listener;
	let applyStyles_action;
	let t1;
	let resizablehandle;
	let div_id_value;
	let div_class_value;
	let div_data_appid_value;
	let div_resize_listener;
	let applyStyles_action_1;
	let div_intro;
	let div_outro;
	let current;
	let mounted;
	let dispose;
	tjsapplicationheader = new TJSApplicationHeader({});
	const if_block_creators = [create_if_block_1$1, create_else_block$1];
	const if_blocks = [];

	function select_block_type_1(ctx, dirty) {
		if (Array.isArray(/*allChildren*/ ctx[12])) return 0;
		return 1;
	}

	current_block_type_index = select_block_type_1(ctx);
	if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);
	resizablehandle = new ResizableHandle({});

	return {
		c() {
			div = element("div");
			create_component(tjsapplicationheader.$$.fragment);
			t0 = space();
			section = element("section");
			if_block.c();
			t1 = space();
			create_component(resizablehandle.$$.fragment);
			attr(section, "class", "window-content");
			add_render_callback(() => /*section_elementresize_handler*/ ctx[21].call(section));
			attr(div, "id", div_id_value = /*application*/ ctx[9].id);
			attr(div, "class", div_class_value = "tjs-app tjs-window-app " + /*application*/ ctx[9].options.classes.join(' '));
			attr(div, "data-appid", div_data_appid_value = /*application*/ ctx[9].appId);
			add_render_callback(() => /*div_elementresize_handler*/ ctx[22].call(div));
		},
		m(target, anchor) {
			insert(target, div, anchor);
			mount_component(tjsapplicationheader, div, null);
			append(div, t0);
			append(div, section);
			if_blocks[current_block_type_index].m(section, null);
			/*section_binding*/ ctx[20](section);
			section_resize_listener = add_resize_listener(section, /*section_elementresize_handler*/ ctx[21].bind(section));
			append(div, t1);
			mount_component(resizablehandle, div, null);
			div_resize_listener = add_resize_listener(div, /*div_elementresize_handler*/ ctx[22].bind(div));
			/*div_binding*/ ctx[23](div);
			current = true;

			if (!mounted) {
				dispose = [
					action_destroyer(applyStyles_action = applyStyles.call(null, section, /*stylesContent*/ ctx[8])),
					listen(div, "pointerdown", /*bringToTop*/ ctx[11], true),
					action_destroyer(applyStyles_action_1 = applyStyles.call(null, div, /*stylesApp*/ ctx[7]))
				];

				mounted = true;
			}
		},
		p(new_ctx, dirty) {
			ctx = new_ctx;
			if_block.p(ctx, dirty);
			if (applyStyles_action && is_function(applyStyles_action.update) && dirty & /*stylesContent*/ 256) applyStyles_action.update.call(null, /*stylesContent*/ ctx[8]);

			if (!current || dirty & /*application*/ 512 && div_id_value !== (div_id_value = /*application*/ ctx[9].id)) {
				attr(div, "id", div_id_value);
			}

			if (!current || dirty & /*application*/ 512 && div_class_value !== (div_class_value = "tjs-app tjs-window-app " + /*application*/ ctx[9].options.classes.join(' '))) {
				attr(div, "class", div_class_value);
			}

			if (!current || dirty & /*application*/ 512 && div_data_appid_value !== (div_data_appid_value = /*application*/ ctx[9].appId)) {
				attr(div, "data-appid", div_data_appid_value);
			}

			if (applyStyles_action_1 && is_function(applyStyles_action_1.update) && dirty & /*stylesApp*/ 128) applyStyles_action_1.update.call(null, /*stylesApp*/ ctx[7]);
		},
		i(local) {
			if (current) return;
			transition_in(tjsapplicationheader.$$.fragment, local);
			transition_in(if_block);
			transition_in(resizablehandle.$$.fragment, local);

			add_render_callback(() => {
				if (div_outro) div_outro.end(1);
				div_intro = create_in_transition(div, /*inTransition*/ ctx[2], /*inTransitionOptions*/ ctx[4]);
				div_intro.start();
			});

			current = true;
		},
		o(local) {
			transition_out(tjsapplicationheader.$$.fragment, local);
			transition_out(if_block);
			transition_out(resizablehandle.$$.fragment, local);
			if (div_intro) div_intro.invalidate();
			div_outro = create_out_transition(div, /*outTransition*/ ctx[3], /*outTransitionOptions*/ ctx[5]);
			current = false;
		},
		d(detaching) {
			if (detaching) detach(div);
			destroy_component(tjsapplicationheader);
			if_blocks[current_block_type_index].d();
			/*section_binding*/ ctx[20](null);
			section_resize_listener();
			destroy_component(resizablehandle);
			div_resize_listener();
			/*div_binding*/ ctx[23](null);
			if (detaching && div_outro) div_outro.end();
			mounted = false;
			run_all(dispose);
		}
	};
}

// (179:11) {:else}
function create_else_block_2(ctx) {
	let current;
	const default_slot_template = /*#slots*/ ctx[19].default;
	const default_slot = create_slot(default_slot_template, ctx, /*$$scope*/ ctx[18], null);

	return {
		c() {
			if (default_slot) default_slot.c();
		},
		m(target, anchor) {
			if (default_slot) {
				default_slot.m(target, anchor);
			}

			current = true;
		},
		p(ctx, dirty) {
			if (default_slot) {
				if (default_slot.p && (!current || dirty & /*$$scope*/ 262144)) {
					update_slot_base(
						default_slot,
						default_slot_template,
						ctx,
						/*$$scope*/ ctx[18],
						!current
						? get_all_dirty_from_scope(/*$$scope*/ ctx[18])
						: get_slot_changes(default_slot_template, /*$$scope*/ ctx[18], dirty, null),
						null
					);
				}
			}
		},
		i(local) {
			if (current) return;
			transition_in(default_slot, local);
			current = true;
		},
		o(local) {
			transition_out(default_slot, local);
			current = false;
		},
		d(detaching) {
			if (default_slot) default_slot.d(detaching);
		}
	};
}

// (177:11) {#if Array.isArray(allChildren)}
function create_if_block_2$1(ctx) {
	let tjscontainer;
	let current;

	tjscontainer = new TJSContainer({
			props: { children: /*allChildren*/ ctx[12] }
		});

	return {
		c() {
			create_component(tjscontainer.$$.fragment);
		},
		m(target, anchor) {
			mount_component(tjscontainer, target, anchor);
			current = true;
		},
		p: noop,
		i(local) {
			if (current) return;
			transition_in(tjscontainer.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(tjscontainer.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(tjscontainer, detaching);
		}
	};
}

// (160:11) {:else}
function create_else_block$1(ctx) {
	let current;
	const default_slot_template = /*#slots*/ ctx[19].default;
	const default_slot = create_slot(default_slot_template, ctx, /*$$scope*/ ctx[18], null);

	return {
		c() {
			if (default_slot) default_slot.c();
		},
		m(target, anchor) {
			if (default_slot) {
				default_slot.m(target, anchor);
			}

			current = true;
		},
		p(ctx, dirty) {
			if (default_slot) {
				if (default_slot.p && (!current || dirty & /*$$scope*/ 262144)) {
					update_slot_base(
						default_slot,
						default_slot_template,
						ctx,
						/*$$scope*/ ctx[18],
						!current
						? get_all_dirty_from_scope(/*$$scope*/ ctx[18])
						: get_slot_changes(default_slot_template, /*$$scope*/ ctx[18], dirty, null),
						null
					);
				}
			}
		},
		i(local) {
			if (current) return;
			transition_in(default_slot, local);
			current = true;
		},
		o(local) {
			transition_out(default_slot, local);
			current = false;
		},
		d(detaching) {
			if (default_slot) default_slot.d(detaching);
		}
	};
}

// (158:11) {#if Array.isArray(allChildren)}
function create_if_block_1$1(ctx) {
	let tjscontainer;
	let current;

	tjscontainer = new TJSContainer({
			props: { children: /*allChildren*/ ctx[12] }
		});

	return {
		c() {
			create_component(tjscontainer.$$.fragment);
		},
		m(target, anchor) {
			mount_component(tjscontainer, target, anchor);
			current = true;
		},
		p: noop,
		i(local) {
			if (current) return;
			transition_in(tjscontainer.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(tjscontainer.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(tjscontainer, detaching);
		}
	};
}

function create_fragment$2(ctx) {
	let current_block_type_index;
	let if_block;
	let if_block_anchor;
	let current;
	const if_block_creators = [create_if_block$2, create_else_block_1];
	const if_blocks = [];

	function select_block_type(ctx, dirty) {
		if (/*bindHeightChanged*/ ctx[10]) return 0;
		return 1;
	}

	current_block_type_index = select_block_type(ctx);
	if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);

	return {
		c() {
			if_block.c();
			if_block_anchor = empty();
		},
		m(target, anchor) {
			if_blocks[current_block_type_index].m(target, anchor);
			insert(target, if_block_anchor, anchor);
			current = true;
		},
		p(ctx, [dirty]) {
			if_block.p(ctx, dirty);
		},
		i(local) {
			if (current) return;
			transition_in(if_block);
			current = true;
		},
		o(local) {
			transition_out(if_block);
			current = false;
		},
		d(detaching) {
			if_blocks[current_block_type_index].d(detaching);
			if (detaching) detach(if_block_anchor);
		}
	};
}

function instance$2($$self, $$props, $$invalidate) {
	let { $$slots: slots = {}, $$scope } = $$props;
	let { elementContent } = $$props;
	let { elementRoot } = $$props;
	let { children = void 0 } = $$props;
	let { heightChanged = false } = $$props;
	let { stylesApp } = $$props;
	let { stylesContent } = $$props;

	// Store the initial `heightChanged` state. If it is truthy then `clientHeight` for the content & root elements
	// are bound to `heightChanged` to signal to any parent component of any change to the client & root.
	const bindHeightChanged = !!heightChanged;

	// If the application is a popOut application then when clicked bring to top. Bound to on pointerdown.
	const bringToTop = () => {
		if (typeof application.options.popOut === 'boolean' && application.options.popOut && application !== ui?.activeWindow) {
			application.bringToTop.call(application);
		}
	};

	// Use a writable store to make `elementContent` and `elementRoot` accessible. A store is used in the case when
	// One root component with an `elementRoot` is replaced with another. Due to timing issues and the onDestroy / outro
	// transitions either of these may be set to null. I will investigate more and file a bug against Svelte.
	if (!getContext('storeElementContent')) {
		setContext('storeElementContent', writable(elementContent));
	}

	if (!getContext('storeElementRoot')) {
		setContext('storeElementRoot', writable(elementRoot));
	}

	const context = getContext('external');

	// Store Foundry Application reference.
	const application = context.application;

	// This component can host multiple children defined via props or in the TyphonJS SvelteData configuration object
	// that are potentially mounted in the content area. If no children defined then this component mounts any slotted
	// child.
	const allChildren = Array.isArray(children)
	? children
	: typeof context === 'object' ? context.children : void 0;

	let { transition = void 0 } = $$props;
	let { inTransition = s_DEFAULT_TRANSITION } = $$props;
	let { outTransition = s_DEFAULT_TRANSITION } = $$props;
	let { transitionOptions = void 0 } = $$props;
	let { inTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS } = $$props;
	let { outTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS } = $$props;

	// Tracks last transition state.
	let oldTransition = void 0;

	let oldTransitionOptions = void 0;

	function section_binding($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementContent = $$value;
			$$invalidate(0, elementContent);
		});
	}

	function section_elementresize_handler() {
		heightChanged = this.clientHeight;
		$$invalidate(6, heightChanged);
	}

	function div_elementresize_handler() {
		heightChanged = this.clientHeight;
		$$invalidate(6, heightChanged);
	}

	function div_binding($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementRoot = $$value;
			$$invalidate(1, elementRoot);
		});
	}

	function section_binding_1($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementContent = $$value;
			$$invalidate(0, elementContent);
		});
	}

	function div_binding_1($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			elementRoot = $$value;
			$$invalidate(1, elementRoot);
		});
	}

	$$self.$$set = $$props => {
		if ('elementContent' in $$props) $$invalidate(0, elementContent = $$props.elementContent);
		if ('elementRoot' in $$props) $$invalidate(1, elementRoot = $$props.elementRoot);
		if ('children' in $$props) $$invalidate(13, children = $$props.children);
		if ('heightChanged' in $$props) $$invalidate(6, heightChanged = $$props.heightChanged);
		if ('stylesApp' in $$props) $$invalidate(7, stylesApp = $$props.stylesApp);
		if ('stylesContent' in $$props) $$invalidate(8, stylesContent = $$props.stylesContent);
		if ('transition' in $$props) $$invalidate(14, transition = $$props.transition);
		if ('inTransition' in $$props) $$invalidate(2, inTransition = $$props.inTransition);
		if ('outTransition' in $$props) $$invalidate(3, outTransition = $$props.outTransition);
		if ('transitionOptions' in $$props) $$invalidate(15, transitionOptions = $$props.transitionOptions);
		if ('inTransitionOptions' in $$props) $$invalidate(4, inTransitionOptions = $$props.inTransitionOptions);
		if ('outTransitionOptions' in $$props) $$invalidate(5, outTransitionOptions = $$props.outTransitionOptions);
		if ('$$scope' in $$props) $$invalidate(18, $$scope = $$props.$$scope);
	};

	$$self.$$.update = () => {
		if ($$self.$$.dirty & /*elementContent*/ 1) {
			// Only update the `elementContent` store if the new `elementContent` is not null or undefined.
			if (elementContent !== void 0 && elementContent !== null) {
				getContext('storeElementContent').set(elementContent);
			}
		}

		if ($$self.$$.dirty & /*elementRoot*/ 2) {
			// Only update the `elementRoot` store if the new `elementRoot` is not null or undefined.
			if (elementRoot !== void 0 && elementRoot !== null) {
				getContext('storeElementRoot').set(elementRoot);
			}
		}

		if ($$self.$$.dirty & /*oldTransition, transition*/ 81920) {
			// Run this reactive block when the last transition state is not equal to the current state.
			if (oldTransition !== transition) {
				// If transition is defined and not the default transition then set it to both in and out transition otherwise
				// set the default transition to both in & out transitions.
				const newTransition = s_DEFAULT_TRANSITION !== transition && typeof transition === 'function'
				? transition
				: s_DEFAULT_TRANSITION;

				$$invalidate(2, inTransition = newTransition);
				$$invalidate(3, outTransition = newTransition);
				$$invalidate(16, oldTransition = newTransition);
			}
		}

		if ($$self.$$.dirty & /*oldTransitionOptions, transitionOptions*/ 163840) {
			// Run this reactive block when the last transition options state is not equal to the current options state.
			if (oldTransitionOptions !== transitionOptions) {
				const newOptions = transitionOptions !== s_DEFAULT_TRANSITION_OPTIONS && typeof transitionOptions === 'object'
				? transitionOptions
				: s_DEFAULT_TRANSITION_OPTIONS;

				$$invalidate(4, inTransitionOptions = newOptions);
				$$invalidate(5, outTransitionOptions = newOptions);
				$$invalidate(17, oldTransitionOptions = newOptions);
			}
		}

		if ($$self.$$.dirty & /*inTransition*/ 4) {
			// Handle cases if inTransition is unset; assign noop default transition function.
			if (typeof inTransition !== 'function') {
				$$invalidate(2, inTransition = s_DEFAULT_TRANSITION);
			}
		}

		if ($$self.$$.dirty & /*outTransition, application*/ 520) {
			{
				// Handle cases if outTransition is unset; assign noop default transition function.
				if (typeof outTransition !== 'function') {
					$$invalidate(3, outTransition = s_DEFAULT_TRANSITION);
				}

				// Set jquery close animation to either run or not when an out transition is changed.
				if (application && typeof application?.options?.defaultCloseAnimation === 'boolean') {
					$$invalidate(9, application.options.defaultCloseAnimation = outTransition === s_DEFAULT_TRANSITION, application);
				}
			}
		}

		if ($$self.$$.dirty & /*inTransitionOptions*/ 16) {
			// Handle cases if inTransitionOptions is unset; assign empty default transition options.
			if (typeof inTransitionOptions !== 'object') {
				$$invalidate(4, inTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS);
			}
		}

		if ($$self.$$.dirty & /*outTransitionOptions*/ 32) {
			// Handle cases if outTransitionOptions is unset; assign empty default transition options.
			if (typeof outTransitionOptions !== 'object') {
				$$invalidate(5, outTransitionOptions = s_DEFAULT_TRANSITION_OPTIONS);
			}
		}
	};

	return [
		elementContent,
		elementRoot,
		inTransition,
		outTransition,
		inTransitionOptions,
		outTransitionOptions,
		heightChanged,
		stylesApp,
		stylesContent,
		application,
		bindHeightChanged,
		bringToTop,
		allChildren,
		children,
		transition,
		transitionOptions,
		oldTransition,
		oldTransitionOptions,
		$$scope,
		slots,
		section_binding,
		section_elementresize_handler,
		div_elementresize_handler,
		div_binding,
		section_binding_1,
		div_binding_1
	];
}

class TJSApplicationShell extends SvelteComponent {
	constructor(options) {
		super();

		init(
			this,
			options,
			instance$2,
			create_fragment$2,
			safe_not_equal,
			{
				elementContent: 0,
				elementRoot: 1,
				children: 13,
				heightChanged: 6,
				stylesApp: 7,
				stylesContent: 8,
				transition: 14,
				inTransition: 2,
				outTransition: 3,
				transitionOptions: 15,
				inTransitionOptions: 4,
				outTransitionOptions: 5
			},
			add_css$1
		);
	}

	get elementContent() {
		return this.$$.ctx[0];
	}

	set elementContent(elementContent) {
		this.$$set({ elementContent });
		flush();
	}

	get elementRoot() {
		return this.$$.ctx[1];
	}

	set elementRoot(elementRoot) {
		this.$$set({ elementRoot });
		flush();
	}

	get children() {
		return this.$$.ctx[13];
	}

	set children(children) {
		this.$$set({ children });
		flush();
	}

	get heightChanged() {
		return this.$$.ctx[6];
	}

	set heightChanged(heightChanged) {
		this.$$set({ heightChanged });
		flush();
	}

	get stylesApp() {
		return this.$$.ctx[7];
	}

	set stylesApp(stylesApp) {
		this.$$set({ stylesApp });
		flush();
	}

	get stylesContent() {
		return this.$$.ctx[8];
	}

	set stylesContent(stylesContent) {
		this.$$set({ stylesContent });
		flush();
	}

	get transition() {
		return this.$$.ctx[14];
	}

	set transition(transition) {
		this.$$set({ transition });
		flush();
	}

	get inTransition() {
		return this.$$.ctx[2];
	}

	set inTransition(inTransition) {
		this.$$set({ inTransition });
		flush();
	}

	get outTransition() {
		return this.$$.ctx[3];
	}

	set outTransition(outTransition) {
		this.$$set({ outTransition });
		flush();
	}

	get transitionOptions() {
		return this.$$.ctx[15];
	}

	set transitionOptions(transitionOptions) {
		this.$$set({ transitionOptions });
		flush();
	}

	get inTransitionOptions() {
		return this.$$.ctx[4];
	}

	set inTransitionOptions(inTransitionOptions) {
		this.$$set({ inTransitionOptions });
		flush();
	}

	get outTransitionOptions() {
		return this.$$.ctx[5];
	}

	set outTransitionOptions(outTransitionOptions) {
		this.$$set({ outTransitionOptions });
		flush();
	}
}

/* src\component\core\dialog\DialogContent.svelte generated by Svelte v3.46.0 */

function add_css(target) {
	append_styles(target, "svelte-14xg9ru", "div.dialog-buttons.svelte-14xg9ru{padding-top:8px}");
}

function get_each_context(ctx, list, i) {
	const child_ctx = ctx.slice();
	child_ctx[15] = list[i];
	return child_ctx;
}

// (202:29) 
function create_if_block_3(ctx) {
	let switch_instance;
	let switch_instance_anchor;
	let current;
	const switch_instance_spread_levels = [/*dialogProps*/ ctx[5]];
	var switch_value = /*dialogComponent*/ ctx[4];

	function switch_props(ctx) {
		let switch_instance_props = {};

		for (let i = 0; i < switch_instance_spread_levels.length; i += 1) {
			switch_instance_props = assign(switch_instance_props, switch_instance_spread_levels[i]);
		}

		return { props: switch_instance_props };
	}

	if (switch_value) {
		switch_instance = new switch_value(switch_props());
		/*switch_instance_binding*/ ctx[12](switch_instance);
	}

	return {
		c() {
			if (switch_instance) create_component(switch_instance.$$.fragment);
			switch_instance_anchor = empty();
		},
		m(target, anchor) {
			if (switch_instance) {
				mount_component(switch_instance, target, anchor);
			}

			insert(target, switch_instance_anchor, anchor);
			current = true;
		},
		p(ctx, dirty) {
			const switch_instance_changes = (dirty & /*dialogProps*/ 32)
			? get_spread_update(switch_instance_spread_levels, [get_spread_object(/*dialogProps*/ ctx[5])])
			: {};

			if (switch_value !== (switch_value = /*dialogComponent*/ ctx[4])) {
				if (switch_instance) {
					group_outros();
					const old_component = switch_instance;

					transition_out(old_component.$$.fragment, 1, 0, () => {
						destroy_component(old_component, 1);
					});

					check_outros();
				}

				if (switch_value) {
					switch_instance = new switch_value(switch_props());
					/*switch_instance_binding*/ ctx[12](switch_instance);
					create_component(switch_instance.$$.fragment);
					transition_in(switch_instance.$$.fragment, 1);
					mount_component(switch_instance, switch_instance_anchor.parentNode, switch_instance_anchor);
				} else {
					switch_instance = null;
				}
			} else if (switch_value) {
				switch_instance.$set(switch_instance_changes);
			}
		},
		i(local) {
			if (current) return;
			if (switch_instance) transition_in(switch_instance.$$.fragment, local);
			current = true;
		},
		o(local) {
			if (switch_instance) transition_out(switch_instance.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			/*switch_instance_binding*/ ctx[12](null);
			if (detaching) detach(switch_instance_anchor);
			if (switch_instance) destroy_component(switch_instance, detaching);
		}
	};
}

// (200:3) {#if typeof content === 'string'}
function create_if_block_2(ctx) {
	let html_tag;
	let html_anchor;

	return {
		c() {
			html_tag = new HtmlTag();
			html_anchor = empty();
			html_tag.a = html_anchor;
		},
		m(target, anchor) {
			html_tag.m(/*content*/ ctx[2], target, anchor);
			insert(target, html_anchor, anchor);
		},
		p(ctx, dirty) {
			if (dirty & /*content*/ 4) html_tag.p(/*content*/ ctx[2]);
		},
		i: noop,
		o: noop,
		d(detaching) {
			if (detaching) detach(html_anchor);
			if (detaching) html_tag.d();
		}
	};
}

// (207:0) {#if buttons.length}
function create_if_block$1(ctx) {
	let div;
	let each_blocks = [];
	let each_1_lookup = new Map();
	let each_value = /*buttons*/ ctx[1];
	const get_key = ctx => /*button*/ ctx[15].id;

	for (let i = 0; i < each_value.length; i += 1) {
		let child_ctx = get_each_context(ctx, each_value, i);
		let key = get_key(child_ctx);
		each_1_lookup.set(key, each_blocks[i] = create_each_block(key, child_ctx));
	}

	return {
		c() {
			div = element("div");

			for (let i = 0; i < each_blocks.length; i += 1) {
				each_blocks[i].c();
			}

			attr(div, "class", "dialog-buttons svelte-14xg9ru");
		},
		m(target, anchor) {
			insert(target, div, anchor);

			for (let i = 0; i < each_blocks.length; i += 1) {
				each_blocks[i].m(div, null);
			}
		},
		p(ctx, dirty) {
			if (dirty & /*buttons, currentButtonId, onClick*/ 74) {
				each_value = /*buttons*/ ctx[1];
				each_blocks = update_keyed_each(each_blocks, dirty, get_key, 1, ctx, each_value, each_1_lookup, div, destroy_block, create_each_block, null, get_each_context);
			}
		},
		d(detaching) {
			if (detaching) detach(div);

			for (let i = 0; i < each_blocks.length; i += 1) {
				each_blocks[i].d();
			}
		}
	};
}

// (214:33) {#if button.icon}
function create_if_block_1(ctx) {
	let html_tag;
	let raw_value = /*button*/ ctx[15].icon + "";
	let html_anchor;

	return {
		c() {
			html_tag = new HtmlTag();
			html_anchor = empty();
			html_tag.a = html_anchor;
		},
		m(target, anchor) {
			html_tag.m(raw_value, target, anchor);
			insert(target, html_anchor, anchor);
		},
		p(ctx, dirty) {
			if (dirty & /*buttons*/ 2 && raw_value !== (raw_value = /*button*/ ctx[15].icon + "")) html_tag.p(raw_value);
		},
		d(detaching) {
			if (detaching) detach(html_anchor);
			if (detaching) html_tag.d();
		}
	};
}

// (209:3) {#each buttons as button (button.id)}
function create_each_block(key_1, ctx) {
	let button;
	let span;
	let t0_value = /*button*/ ctx[15].label + "";
	let t0;
	let span_title_value;
	let t1;
	let button_class_value;
	let applyStyles_action;
	let mounted;
	let dispose;
	let if_block = /*button*/ ctx[15].icon && create_if_block_1(ctx);

	function click_handler() {
		return /*click_handler*/ ctx[13](/*button*/ ctx[15]);
	}

	return {
		key: key_1,
		first: null,
		c() {
			button = element("button");
			span = element("span");
			if (if_block) if_block.c();
			t0 = text(t0_value);
			t1 = space();
			attr(span, "title", span_title_value = /*button*/ ctx[15].title);
			attr(button, "class", button_class_value = "dialog-button " + /*button*/ ctx[15].id);
			toggle_class(button, "default", /*button*/ ctx[15].id === /*currentButtonId*/ ctx[3]);
			this.first = button;
		},
		m(target, anchor) {
			insert(target, button, anchor);
			append(button, span);
			if (if_block) if_block.m(span, null);
			append(span, t0);
			append(button, t1);

			if (!mounted) {
				dispose = [
					listen(button, "click", click_handler),
					action_destroyer(applyStyles_action = applyStyles.call(null, button, /*button*/ ctx[15].styles))
				];

				mounted = true;
			}
		},
		p(new_ctx, dirty) {
			ctx = new_ctx;

			if (/*button*/ ctx[15].icon) {
				if (if_block) {
					if_block.p(ctx, dirty);
				} else {
					if_block = create_if_block_1(ctx);
					if_block.c();
					if_block.m(span, t0);
				}
			} else if (if_block) {
				if_block.d(1);
				if_block = null;
			}

			if (dirty & /*buttons*/ 2 && t0_value !== (t0_value = /*button*/ ctx[15].label + "")) set_data(t0, t0_value);

			if (dirty & /*buttons*/ 2 && span_title_value !== (span_title_value = /*button*/ ctx[15].title)) {
				attr(span, "title", span_title_value);
			}

			if (dirty & /*buttons*/ 2 && button_class_value !== (button_class_value = "dialog-button " + /*button*/ ctx[15].id)) {
				attr(button, "class", button_class_value);
			}

			if (applyStyles_action && is_function(applyStyles_action.update) && dirty & /*buttons*/ 2) applyStyles_action.update.call(null, /*button*/ ctx[15].styles);

			if (dirty & /*buttons, buttons, currentButtonId*/ 10) {
				toggle_class(button, "default", /*button*/ ctx[15].id === /*currentButtonId*/ ctx[3]);
			}
		},
		d(detaching) {
			if (detaching) detach(button);
			if (if_block) if_block.d();
			mounted = false;
			run_all(dispose);
		}
	};
}

function create_fragment$1(ctx) {
	let t0;
	let div;
	let current_block_type_index;
	let if_block0;
	let t1;
	let if_block1_anchor;
	let current;
	let mounted;
	let dispose;
	const if_block_creators = [create_if_block_2, create_if_block_3];
	const if_blocks = [];

	function select_block_type(ctx, dirty) {
		if (typeof /*content*/ ctx[2] === 'string') return 0;
		if (/*dialogComponent*/ ctx[4]) return 1;
		return -1;
	}

	if (~(current_block_type_index = select_block_type(ctx))) {
		if_block0 = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);
	}

	let if_block1 = /*buttons*/ ctx[1].length && create_if_block$1(ctx);

	return {
		c() {
			t0 = space();
			div = element("div");
			if (if_block0) if_block0.c();
			t1 = space();
			if (if_block1) if_block1.c();
			if_block1_anchor = empty();
			attr(div, "class", "dialog-content");
		},
		m(target, anchor) {
			insert(target, t0, anchor);
			insert(target, div, anchor);

			if (~current_block_type_index) {
				if_blocks[current_block_type_index].m(div, null);
			}

			insert(target, t1, anchor);
			if (if_block1) if_block1.m(target, anchor);
			insert(target, if_block1_anchor, anchor);
			current = true;

			if (!mounted) {
				dispose = listen(document.body, "keydown", /*onKeydown*/ ctx[7]);
				mounted = true;
			}
		},
		p(ctx, [dirty]) {
			let previous_block_index = current_block_type_index;
			current_block_type_index = select_block_type(ctx);

			if (current_block_type_index === previous_block_index) {
				if (~current_block_type_index) {
					if_blocks[current_block_type_index].p(ctx, dirty);
				}
			} else {
				if (if_block0) {
					group_outros();

					transition_out(if_blocks[previous_block_index], 1, 1, () => {
						if_blocks[previous_block_index] = null;
					});

					check_outros();
				}

				if (~current_block_type_index) {
					if_block0 = if_blocks[current_block_type_index];

					if (!if_block0) {
						if_block0 = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);
						if_block0.c();
					} else {
						if_block0.p(ctx, dirty);
					}

					transition_in(if_block0, 1);
					if_block0.m(div, null);
				} else {
					if_block0 = null;
				}
			}

			if (/*buttons*/ ctx[1].length) {
				if (if_block1) {
					if_block1.p(ctx, dirty);
				} else {
					if_block1 = create_if_block$1(ctx);
					if_block1.c();
					if_block1.m(if_block1_anchor.parentNode, if_block1_anchor);
				}
			} else if (if_block1) {
				if_block1.d(1);
				if_block1 = null;
			}
		},
		i(local) {
			if (current) return;
			transition_in(if_block0);
			current = true;
		},
		o(local) {
			transition_out(if_block0);
			current = false;
		},
		d(detaching) {
			if (detaching) detach(t0);
			if (detaching) detach(div);

			if (~current_block_type_index) {
				if_blocks[current_block_type_index].d();
			}

			if (detaching) detach(t1);
			if (if_block1) if_block1.d(detaching);
			if (detaching) detach(if_block1_anchor);
			mounted = false;
			dispose();
		}
	};
}

const s_REGEX_HTML = /^\s*<.*>$/;

function instance$1($$self, $$props, $$invalidate) {
	let { data = {} } = $$props;
	let { autoClose = true } = $$props;
	let { preventDefault = false } = $$props;
	let { stopPropagation = false } = $$props;
	let { dialogInstance = void 0 } = $$props;
	let buttons;
	let content = void 0;
	let dialogComponent;
	let dialogProps = {};
	let application = getContext('external').application;
	let currentButtonId = data.default;

	async function onClick(button) {
		try {
			let result = null;

			// Accept either callback or onclick as the function / data to invoke.
			const invoke = button.callback ?? button.onclick;

			switch (typeof invoke) {
				case 'function':
					// Passing back the HTML element is to keep with the existing Foundry API, however second parameter is
					// the Svelte component instance.
					result = await invoke(
						application.options.jQuery
						? application.element
						: application.element[0],
						dialogInstance
					);
					break;
				case 'string':
					// Attempt lookup by function name in dialog instance component.
					if (dialogInstance !== void 0 && typeof dialogInstance[invoke] === 'function') {
						result = await dialogInstance[invoke](
							application.options.jQuery
							? application.element
							: application.element[0],
							dialogInstance
						);
					}
					break;
			}

			// Delay closing to next clock tick to be able to return result.
			if (autoClose) {
				setTimeout(() => application.close(), 0);
			}

			return result;
		} catch(err) {
			ui.notifications.error(err);
			throw new Error(err);
		}
	}

	function onKeydown(event) {
		/**
 * If this dialog is not the activeWindow then return immediately. See {@link SvelteApplication.bringToTop} as
 * SvelteApplication overrides core Foundry and always sets the activeWindow when `bringToTop` is invoked.
 */
		if (event.key !== 'Escape' && ui.activeWindow !== application) {
			return;
		}

		switch (event.key) {
			case 'ArrowLeft':
				{
					event.preventDefault();
					event.stopPropagation();
					const currentIndex = buttons.findIndex(button => button.id === currentButtonId);

					if (buttons.length && currentIndex > 0) {
						$$invalidate(3, currentButtonId = buttons[currentIndex - 1].id);
					}

					break;
				}
			case 'ArrowRight':
				{
					event.preventDefault();
					event.stopPropagation();
					const currentIndex = buttons.findIndex(button => button.id === currentButtonId);

					if (buttons.length && currentIndex < buttons.length - 1) {
						$$invalidate(3, currentButtonId = buttons[currentIndex + 1].id);
					}

					break;
				}
			case 'Escape':
				event.preventDefault();
				event.stopPropagation();
				return application.close();
			case 'Enter':
				event.preventDefault();
				event.stopPropagation();
				if (currentButtonId && isObject(data.buttons) && currentButtonId in data.buttons) {
					onClick(data.buttons[currentButtonId]);
				}
				break;
			default:
				if (preventDefault) {
					event.preventDefault();
				}
				if (stopPropagation) {
					event.stopPropagation();
				}
				break;
		}
	}

	function switch_instance_binding($$value) {
		binding_callbacks[$$value ? 'unshift' : 'push'](() => {
			dialogInstance = $$value;
			$$invalidate(0, dialogInstance);
		});
	}

	const click_handler = button => onClick(button);

	$$self.$$set = $$props => {
		if ('data' in $$props) $$invalidate(8, data = $$props.data);
		if ('autoClose' in $$props) $$invalidate(9, autoClose = $$props.autoClose);
		if ('preventDefault' in $$props) $$invalidate(10, preventDefault = $$props.preventDefault);
		if ('stopPropagation' in $$props) $$invalidate(11, stopPropagation = $$props.stopPropagation);
		if ('dialogInstance' in $$props) $$invalidate(0, dialogInstance = $$props.dialogInstance);
	};

	$$self.$$.update = () => {
		if ($$self.$$.dirty & /*data*/ 256) {
			// If `data.buttons` is not an object then set an empty array otherwise reduce the button data.
			{
				$$invalidate(1, buttons = !isObject(data.buttons)
				? []
				: Object.keys(data.buttons).reduce(
						(array, key) => {
							const b = data.buttons[key];

							// Handle icon and treat bare strings as the icon class; otherwise assume the icon is fully formed HTML.
							const icon = typeof b.icon !== 'string'
							? void 0
							: s_REGEX_HTML.test(b.icon)
								? b.icon
								: `<i class="${b.icon}"></i>`;

							const label = typeof b.label === 'string'
							? `${icon !== void 0 ? ' ' : ''}${localize(b.label)}`
							: '';

							const title = typeof b.title === 'string' ? localize(b.title) : void 0;

							// Test any condition supplied otherwise default to true.
							const condition = typeof b.condition === 'function'
							? b.condition.call(b)
							: b.condition ?? true;

							if (condition) {
								array.push({ ...b, id: key, icon, label, title });
							}

							return array;
						},
						[]
					));
			}
		}

		if ($$self.$$.dirty & /*buttons, currentButtonId*/ 10) {
			/**
 * This reactivity block will trigger on arrow left / right key presses _and_ when buttons change. It is OK for it to
 * trigger on both.
 */
			if (!buttons.find(button => button.id === currentButtonId)) {
				$$invalidate(3, currentButtonId = void 0);
			}
		}

		if ($$self.$$.dirty & /*content, data*/ 260) {
			if (content !== data.content) {
				$$invalidate(2, content = data.content); // Only update the content if it has changed.

				try {
					if (isSvelteComponent(content)) {
						$$invalidate(4, dialogComponent = content);
						$$invalidate(5, dialogProps = {});
					} else if (typeof content === 'object') {
						const svelteConfig = parseSvelteConfig(content, application);
						$$invalidate(4, dialogComponent = svelteConfig.class);
						$$invalidate(5, dialogProps = svelteConfig.props ?? {});

						// Check for any children parsed and added to the external context.
						const children = svelteConfig?.context?.get('external')?.children;

						// If so add to dialogProps.
						if (Array.isArray(children)) {
							$$invalidate(5, dialogProps.children = children, dialogProps);
						}
					} else {
						$$invalidate(4, dialogComponent = void 0);
						$$invalidate(5, dialogProps = {});
					}
				} catch(err) {
					$$invalidate(4, dialogComponent = void 0);
					$$invalidate(5, dialogProps = {});
					$$invalidate(2, content = err.message);
					console.error(err);
				}
			}
		}
	};

	return [
		dialogInstance,
		buttons,
		content,
		currentButtonId,
		dialogComponent,
		dialogProps,
		onClick,
		onKeydown,
		data,
		autoClose,
		preventDefault,
		stopPropagation,
		switch_instance_binding,
		click_handler
	];
}

class DialogContent extends SvelteComponent {
	constructor(options) {
		super();

		init(
			this,
			options,
			instance$1,
			create_fragment$1,
			safe_not_equal,
			{
				data: 8,
				autoClose: 9,
				preventDefault: 10,
				stopPropagation: 11,
				dialogInstance: 0
			},
			add_css
		);
	}
}

/* src\component\core\dialog\DialogShell.svelte generated by Svelte v3.46.0 */

function create_else_block(ctx) {
	let applicationshell;
	let updating_elementRoot;
	let updating_elementContent;
	let current;
	const applicationshell_spread_levels = [/*appProps*/ ctx[6]];

	function applicationshell_elementRoot_binding_1(value) {
		/*applicationshell_elementRoot_binding_1*/ ctx[16](value);
	}

	function applicationshell_elementContent_binding_1(value) {
		/*applicationshell_elementContent_binding_1*/ ctx[17](value);
	}

	let applicationshell_props = {
		$$slots: { default: [create_default_slot_2] },
		$$scope: { ctx }
	};

	for (let i = 0; i < applicationshell_spread_levels.length; i += 1) {
		applicationshell_props = assign(applicationshell_props, applicationshell_spread_levels[i]);
	}

	if (/*elementRoot*/ ctx[1] !== void 0) {
		applicationshell_props.elementRoot = /*elementRoot*/ ctx[1];
	}

	if (/*elementContent*/ ctx[0] !== void 0) {
		applicationshell_props.elementContent = /*elementContent*/ ctx[0];
	}

	applicationshell = new ApplicationShell({ props: applicationshell_props });
	binding_callbacks.push(() => bind(applicationshell, 'elementRoot', applicationshell_elementRoot_binding_1));
	binding_callbacks.push(() => bind(applicationshell, 'elementContent', applicationshell_elementContent_binding_1));

	return {
		c() {
			create_component(applicationshell.$$.fragment);
		},
		m(target, anchor) {
			mount_component(applicationshell, target, anchor);
			current = true;
		},
		p(ctx, dirty) {
			const applicationshell_changes = (dirty & /*appProps*/ 64)
			? get_spread_update(applicationshell_spread_levels, [get_spread_object(/*appProps*/ ctx[6])])
			: {};

			if (dirty & /*$$scope, data, autoClose, dialogComponent*/ 1049100) {
				applicationshell_changes.$$scope = { dirty, ctx };
			}

			if (!updating_elementRoot && dirty & /*elementRoot*/ 2) {
				updating_elementRoot = true;
				applicationshell_changes.elementRoot = /*elementRoot*/ ctx[1];
				add_flush_callback(() => updating_elementRoot = false);
			}

			if (!updating_elementContent && dirty & /*elementContent*/ 1) {
				updating_elementContent = true;
				applicationshell_changes.elementContent = /*elementContent*/ ctx[0];
				add_flush_callback(() => updating_elementContent = false);
			}

			applicationshell.$set(applicationshell_changes);
		},
		i(local) {
			if (current) return;
			transition_in(applicationshell.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(applicationshell.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(applicationshell, detaching);
		}
	};
}

// (180:0) {#if modal}
function create_if_block(ctx) {
	let tjsglasspane;
	let current;

	const tjsglasspane_spread_levels = [
		{
			id: `${/*application*/ ctx[4].id}-glasspane`
		},
		{ preventDefault: false },
		{ stopPropagation: false },
		/*modalProps*/ ctx[7],
		{ zIndex: /*zIndex*/ ctx[8] }
	];

	let tjsglasspane_props = {
		$$slots: { default: [create_default_slot] },
		$$scope: { ctx }
	};

	for (let i = 0; i < tjsglasspane_spread_levels.length; i += 1) {
		tjsglasspane_props = assign(tjsglasspane_props, tjsglasspane_spread_levels[i]);
	}

	tjsglasspane = new TJSGlassPane({ props: tjsglasspane_props });

	return {
		c() {
			create_component(tjsglasspane.$$.fragment);
		},
		m(target, anchor) {
			mount_component(tjsglasspane, target, anchor);
			current = true;
		},
		p(ctx, dirty) {
			const tjsglasspane_changes = (dirty & /*application, modalProps, zIndex*/ 400)
			? get_spread_update(tjsglasspane_spread_levels, [
					dirty & /*application*/ 16 && {
						id: `${/*application*/ ctx[4].id}-glasspane`
					},
					tjsglasspane_spread_levels[1],
					tjsglasspane_spread_levels[2],
					dirty & /*modalProps*/ 128 && get_spread_object(/*modalProps*/ ctx[7]),
					dirty & /*zIndex*/ 256 && { zIndex: /*zIndex*/ ctx[8] }
				])
			: {};

			if (dirty & /*$$scope, appProps, elementRoot, elementContent, data, autoClose, dialogComponent*/ 1049167) {
				tjsglasspane_changes.$$scope = { dirty, ctx };
			}

			tjsglasspane.$set(tjsglasspane_changes);
		},
		i(local) {
			if (current) return;
			transition_in(tjsglasspane.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(tjsglasspane.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(tjsglasspane, detaching);
		}
	};
}

// (187:3) <ApplicationShell bind:elementRoot bind:elementContent {...appProps}>
function create_default_slot_2(ctx) {
	let dialogcontent;
	let updating_autoClose;
	let updating_dialogInstance;
	let current;

	function dialogcontent_autoClose_binding_1(value) {
		/*dialogcontent_autoClose_binding_1*/ ctx[14](value);
	}

	function dialogcontent_dialogInstance_binding_1(value) {
		/*dialogcontent_dialogInstance_binding_1*/ ctx[15](value);
	}

	let dialogcontent_props = { data: /*data*/ ctx[3] };

	if (/*autoClose*/ ctx[9] !== void 0) {
		dialogcontent_props.autoClose = /*autoClose*/ ctx[9];
	}

	if (/*dialogComponent*/ ctx[2] !== void 0) {
		dialogcontent_props.dialogInstance = /*dialogComponent*/ ctx[2];
	}

	dialogcontent = new DialogContent({ props: dialogcontent_props });
	binding_callbacks.push(() => bind(dialogcontent, 'autoClose', dialogcontent_autoClose_binding_1));
	binding_callbacks.push(() => bind(dialogcontent, 'dialogInstance', dialogcontent_dialogInstance_binding_1));

	return {
		c() {
			create_component(dialogcontent.$$.fragment);
		},
		m(target, anchor) {
			mount_component(dialogcontent, target, anchor);
			current = true;
		},
		p(ctx, dirty) {
			const dialogcontent_changes = {};
			if (dirty & /*data*/ 8) dialogcontent_changes.data = /*data*/ ctx[3];

			if (!updating_autoClose && dirty & /*autoClose*/ 512) {
				updating_autoClose = true;
				dialogcontent_changes.autoClose = /*autoClose*/ ctx[9];
				add_flush_callback(() => updating_autoClose = false);
			}

			if (!updating_dialogInstance && dirty & /*dialogComponent*/ 4) {
				updating_dialogInstance = true;
				dialogcontent_changes.dialogInstance = /*dialogComponent*/ ctx[2];
				add_flush_callback(() => updating_dialogInstance = false);
			}

			dialogcontent.$set(dialogcontent_changes);
		},
		i(local) {
			if (current) return;
			transition_in(dialogcontent.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(dialogcontent.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(dialogcontent, detaching);
		}
	};
}

// (182:6) <ApplicationShell bind:elementRoot bind:elementContent {...appProps}>
function create_default_slot_1(ctx) {
	let dialogcontent;
	let updating_autoClose;
	let updating_dialogInstance;
	let current;

	function dialogcontent_autoClose_binding(value) {
		/*dialogcontent_autoClose_binding*/ ctx[10](value);
	}

	function dialogcontent_dialogInstance_binding(value) {
		/*dialogcontent_dialogInstance_binding*/ ctx[11](value);
	}

	let dialogcontent_props = {
		stopPropagation: true,
		data: /*data*/ ctx[3]
	};

	if (/*autoClose*/ ctx[9] !== void 0) {
		dialogcontent_props.autoClose = /*autoClose*/ ctx[9];
	}

	if (/*dialogComponent*/ ctx[2] !== void 0) {
		dialogcontent_props.dialogInstance = /*dialogComponent*/ ctx[2];
	}

	dialogcontent = new DialogContent({ props: dialogcontent_props });
	binding_callbacks.push(() => bind(dialogcontent, 'autoClose', dialogcontent_autoClose_binding));
	binding_callbacks.push(() => bind(dialogcontent, 'dialogInstance', dialogcontent_dialogInstance_binding));

	return {
		c() {
			create_component(dialogcontent.$$.fragment);
		},
		m(target, anchor) {
			mount_component(dialogcontent, target, anchor);
			current = true;
		},
		p(ctx, dirty) {
			const dialogcontent_changes = {};
			if (dirty & /*data*/ 8) dialogcontent_changes.data = /*data*/ ctx[3];

			if (!updating_autoClose && dirty & /*autoClose*/ 512) {
				updating_autoClose = true;
				dialogcontent_changes.autoClose = /*autoClose*/ ctx[9];
				add_flush_callback(() => updating_autoClose = false);
			}

			if (!updating_dialogInstance && dirty & /*dialogComponent*/ 4) {
				updating_dialogInstance = true;
				dialogcontent_changes.dialogInstance = /*dialogComponent*/ ctx[2];
				add_flush_callback(() => updating_dialogInstance = false);
			}

			dialogcontent.$set(dialogcontent_changes);
		},
		i(local) {
			if (current) return;
			transition_in(dialogcontent.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(dialogcontent.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(dialogcontent, detaching);
		}
	};
}

// (181:3) <TJSGlassPane id={`${application.id}-glasspane`} preventDefault={false} stopPropagation={false} {...modalProps} {zIndex}>
function create_default_slot(ctx) {
	let applicationshell;
	let updating_elementRoot;
	let updating_elementContent;
	let current;
	const applicationshell_spread_levels = [/*appProps*/ ctx[6]];

	function applicationshell_elementRoot_binding(value) {
		/*applicationshell_elementRoot_binding*/ ctx[12](value);
	}

	function applicationshell_elementContent_binding(value) {
		/*applicationshell_elementContent_binding*/ ctx[13](value);
	}

	let applicationshell_props = {
		$$slots: { default: [create_default_slot_1] },
		$$scope: { ctx }
	};

	for (let i = 0; i < applicationshell_spread_levels.length; i += 1) {
		applicationshell_props = assign(applicationshell_props, applicationshell_spread_levels[i]);
	}

	if (/*elementRoot*/ ctx[1] !== void 0) {
		applicationshell_props.elementRoot = /*elementRoot*/ ctx[1];
	}

	if (/*elementContent*/ ctx[0] !== void 0) {
		applicationshell_props.elementContent = /*elementContent*/ ctx[0];
	}

	applicationshell = new ApplicationShell({ props: applicationshell_props });
	binding_callbacks.push(() => bind(applicationshell, 'elementRoot', applicationshell_elementRoot_binding));
	binding_callbacks.push(() => bind(applicationshell, 'elementContent', applicationshell_elementContent_binding));

	return {
		c() {
			create_component(applicationshell.$$.fragment);
		},
		m(target, anchor) {
			mount_component(applicationshell, target, anchor);
			current = true;
		},
		p(ctx, dirty) {
			const applicationshell_changes = (dirty & /*appProps*/ 64)
			? get_spread_update(applicationshell_spread_levels, [get_spread_object(/*appProps*/ ctx[6])])
			: {};

			if (dirty & /*$$scope, data, autoClose, dialogComponent*/ 1049100) {
				applicationshell_changes.$$scope = { dirty, ctx };
			}

			if (!updating_elementRoot && dirty & /*elementRoot*/ 2) {
				updating_elementRoot = true;
				applicationshell_changes.elementRoot = /*elementRoot*/ ctx[1];
				add_flush_callback(() => updating_elementRoot = false);
			}

			if (!updating_elementContent && dirty & /*elementContent*/ 1) {
				updating_elementContent = true;
				applicationshell_changes.elementContent = /*elementContent*/ ctx[0];
				add_flush_callback(() => updating_elementContent = false);
			}

			applicationshell.$set(applicationshell_changes);
		},
		i(local) {
			if (current) return;
			transition_in(applicationshell.$$.fragment, local);
			current = true;
		},
		o(local) {
			transition_out(applicationshell.$$.fragment, local);
			current = false;
		},
		d(detaching) {
			destroy_component(applicationshell, detaching);
		}
	};
}

function create_fragment(ctx) {
	let current_block_type_index;
	let if_block;
	let if_block_anchor;
	let current;
	const if_block_creators = [create_if_block, create_else_block];
	const if_blocks = [];

	function select_block_type(ctx, dirty) {
		if (/*modal*/ ctx[5]) return 0;
		return 1;
	}

	current_block_type_index = select_block_type(ctx);
	if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);

	return {
		c() {
			if_block.c();
			if_block_anchor = empty();
		},
		m(target, anchor) {
			if_blocks[current_block_type_index].m(target, anchor);
			insert(target, if_block_anchor, anchor);
			current = true;
		},
		p(ctx, [dirty]) {
			let previous_block_index = current_block_type_index;
			current_block_type_index = select_block_type(ctx);

			if (current_block_type_index === previous_block_index) {
				if_blocks[current_block_type_index].p(ctx, dirty);
			} else {
				group_outros();

				transition_out(if_blocks[previous_block_index], 1, 1, () => {
					if_blocks[previous_block_index] = null;
				});

				check_outros();
				if_block = if_blocks[current_block_type_index];

				if (!if_block) {
					if_block = if_blocks[current_block_type_index] = if_block_creators[current_block_type_index](ctx);
					if_block.c();
				} else {
					if_block.p(ctx, dirty);
				}

				transition_in(if_block, 1);
				if_block.m(if_block_anchor.parentNode, if_block_anchor);
			}
		},
		i(local) {
			if (current) return;
			transition_in(if_block);
			current = true;
		},
		o(local) {
			transition_out(if_block);
			current = false;
		},
		d(detaching) {
			if_blocks[current_block_type_index].d(detaching);
			if (detaching) detach(if_block_anchor);
		}
	};
}

const s_MODAL_BACKGROUND = '#50505080';

function instance($$self, $$props, $$invalidate) {
	let { elementContent } = $$props;
	let { elementRoot } = $$props;
	let { data = {} } = $$props;
	let { dialogComponent = void 0 } = $$props;
	const application = getContext('external').application;
	const s_MODAL_TRANSITION = fade;
	const s_MODAL_TRANSITION_OPTIONS = { duration: 200 };
	let modal = void 0;

	// Stores props for the ApplicationShell.
	const appProps = {
		// Stores any transition functions.
		transition: void 0,
		inTransition: void 0,
		outTransition: void 0,
		// Stores properties to set for options for any transitions.
		transitionOptions: void 0,
		inTransitionOptions: void 0,
		outTransitionOptions: void 0,
		// Stores any style overrides for application shell.
		stylesApp: void 0,
		stylesContent: void 0
	};

	const modalProps = {
		// Background CSS style string.
		background: void 0,
		// Stores any transition functions.
		transition: void 0,
		inTransition: void 0,
		outTransition: void 0,
		// Stores properties to set for options for any transitions.
		transitionOptions: void 0,
		inTransitionOptions: void 0,
		outTransitionOptions: void 0
	};

	let zIndex = void 0;

	// Automatically close the dialog on button click handler completion.
	let autoClose = true;

	// Only set modal once on mount. You can't change between a modal an non-modal dialog during runtime.
	if (modal === void 0) {
		modal = typeof data?.modal === 'boolean' ? data.modal : false;
	}

	function dialogcontent_autoClose_binding(value) {
		autoClose = value;
		(((($$invalidate(9, autoClose), $$invalidate(3, data)), $$invalidate(5, modal)), $$invalidate(8, zIndex)), $$invalidate(4, application));
	}

	function dialogcontent_dialogInstance_binding(value) {
		dialogComponent = value;
		$$invalidate(2, dialogComponent);
	}

	function applicationshell_elementRoot_binding(value) {
		elementRoot = value;
		$$invalidate(1, elementRoot);
	}

	function applicationshell_elementContent_binding(value) {
		elementContent = value;
		$$invalidate(0, elementContent);
	}

	function dialogcontent_autoClose_binding_1(value) {
		autoClose = value;
		(((($$invalidate(9, autoClose), $$invalidate(3, data)), $$invalidate(5, modal)), $$invalidate(8, zIndex)), $$invalidate(4, application));
	}

	function dialogcontent_dialogInstance_binding_1(value) {
		dialogComponent = value;
		$$invalidate(2, dialogComponent);
	}

	function applicationshell_elementRoot_binding_1(value) {
		elementRoot = value;
		$$invalidate(1, elementRoot);
	}

	function applicationshell_elementContent_binding_1(value) {
		elementContent = value;
		$$invalidate(0, elementContent);
	}

	$$self.$$set = $$props => {
		if ('elementContent' in $$props) $$invalidate(0, elementContent = $$props.elementContent);
		if ('elementRoot' in $$props) $$invalidate(1, elementRoot = $$props.elementRoot);
		if ('data' in $$props) $$invalidate(3, data = $$props.data);
		if ('dialogComponent' in $$props) $$invalidate(2, dialogComponent = $$props.dialogComponent);
	};

	$$self.$$.update = () => {
		if ($$self.$$.dirty & /*data, modal, zIndex, application*/ 312) {
			// Retrieve values from the DialogData object and also potentially set any SvelteApplication accessors.
			// Explicit checks are performed against existing local variables as the only externally reactive variable is `data`.
			// All of the checks below trigger when there are any external changes to the `data` prop.
			// Prevent any unnecessary changing of local & `application` variables unless actual changes occur.
			// Foundry App options --------------------------------------------------------------------------------------------
			if (typeof data === 'object') {
				$$invalidate(9, autoClose = typeof data.autoClose === 'boolean'
				? data.autoClose
				: true);

				const newZIndex = Number.isInteger(data.zIndex) || data.zIndex === null
				? data.zIndex
				: modal
					? Number.MAX_SAFE_INTEGER
					: Number.MAX_SAFE_INTEGER - 1;

				if (zIndex !== newZIndex) {
					$$invalidate(8, zIndex = newZIndex);
				}

				// Update the main foundry options when data changes. Perform explicit checks against existing data in `application`.
				const newDraggable = data.draggable ?? true;

				if (application.reactive.draggable !== newDraggable) {
					$$invalidate(4, application.reactive.draggable = newDraggable, application);
				}

				const newPopOut = data.popOut ?? true;

				if (application.reactive.popOut !== newPopOut) {
					$$invalidate(4, application.reactive.popOut = newPopOut, application);
				}

				const newResizable = data.resizable ?? false;

				if (application.reactive.resizable !== newResizable) {
					$$invalidate(4, application.reactive.resizable = newResizable, application);
				}

				// Note application.title from Application localizes `options.title`, so compare with `application.options.title`.
				const newTitle = data.title ?? 'Dialog';

				if (newTitle !== application?.options?.title) {
					$$invalidate(4, application.reactive.title = newTitle, application);
				}

				if (application.position.zIndex !== zIndex) {
					$$invalidate(4, application.position.zIndex = zIndex, application);
				}
			}
		}

		if ($$self.$$.dirty & /*data, appProps*/ 72) {
			// ApplicationShell transition options ----------------------------------------------------------------------------
			if (typeof data?.transition === 'object') {
				// Store data.transitions to shorten statements below.
				const d = data.transition;

				if (d?.transition !== appProps.transition) {
					$$invalidate(6, appProps.transition = d.transition, appProps);
				}

				if (d?.inTransition !== appProps.inTransition) {
					$$invalidate(6, appProps.inTransition = d.inTransition, appProps);
				}

				if (d?.outTransition !== appProps.outTransition) {
					$$invalidate(6, appProps.outTransition = d.outTransition, appProps);
				}

				if (d?.transitionOptions !== appProps.transitionOptions) {
					$$invalidate(6, appProps.transitionOptions = d.transitionOptions, appProps);
				}

				if (d?.inTransitionOptions !== appProps.inTransitionOptions) {
					$$invalidate(6, appProps.inTransitionOptions = d.inTransitionOptions, appProps);
				}

				if (d?.outTransitionOptions !== appProps.outTransitionOptions) {
					$$invalidate(6, appProps.outTransitionOptions = d.outTransitionOptions, appProps);
				}
			}
		}

		if ($$self.$$.dirty & /*data, modalProps*/ 136) {
			// Modal options --------------------------------------------------------------------------------------------------
			{
				const newModalBackground = typeof data?.modalOptions?.background === 'string'
				? data.modalOptions.background
				: s_MODAL_BACKGROUND;

				if (newModalBackground !== modalProps.background) {
					$$invalidate(7, modalProps.background = newModalBackground, modalProps);
				}
			}
		}

		if ($$self.$$.dirty & /*data, modalProps*/ 136) {
			if (typeof data?.modalOptions?.transition === 'object') {
				// Store data.transitions to shorten statements below.
				const d = data.modalOptions.transition;

				if (d?.transition !== modalProps.transition) {
					$$invalidate(
						7,
						modalProps.transition = typeof d?.transition === 'function'
						? d.transition
						: s_MODAL_TRANSITION,
						modalProps
					);
				}

				if (d?.inTransition !== modalProps.inTransition) {
					$$invalidate(7, modalProps.inTransition = d.inTransition, modalProps);
				}

				if (d?.outTransition !== modalProps.outTransition) {
					$$invalidate(7, modalProps.outTransition = d.outTransition, modalProps);
				}

				// Provide default transition options if not defined.
				if (d?.transitionOptions !== modalProps.transitionOptions) {
					$$invalidate(
						7,
						modalProps.transitionOptions = typeof d?.transitionOptions === 'object'
						? d.transitionOptions
						: s_MODAL_TRANSITION_OPTIONS,
						modalProps
					);
				}

				if (d?.inTransitionOptions !== modalProps.inTransitionOptions) {
					$$invalidate(7, modalProps.inTransitionOptions = d.inTransitionOptions, modalProps);
				}

				if (d?.outTransitionOptions !== modalProps.outTransitionOptions) {
					$$invalidate(7, modalProps.outTransitionOptions = d.outTransitionOptions, modalProps);
				}
			} else // Provide a fallback / default glass pane transition when `data.modalOptions.transition` is not defined.
			{
				const newModalTransition = typeof data?.modalOptions?.transition?.transition === 'function'
				? data.modalOptions.transition.transition
				: s_MODAL_TRANSITION;

				if (newModalTransition !== modalProps.transition) {
					$$invalidate(7, modalProps.transition = newModalTransition, modalProps);
				}

				const newModalTransitionOptions = typeof data?.modalOptions?.transitionOptions === 'object'
				? data.modalOptions.transitionOptions
				: s_MODAL_TRANSITION_OPTIONS;

				if (newModalTransitionOptions !== modalProps.transitionOptions) {
					$$invalidate(7, modalProps.transitionOptions = newModalTransitionOptions, modalProps);
				}
			}
		}
	};

	return [
		elementContent,
		elementRoot,
		dialogComponent,
		data,
		application,
		modal,
		appProps,
		modalProps,
		zIndex,
		autoClose,
		dialogcontent_autoClose_binding,
		dialogcontent_dialogInstance_binding,
		applicationshell_elementRoot_binding,
		applicationshell_elementContent_binding,
		dialogcontent_autoClose_binding_1,
		dialogcontent_dialogInstance_binding_1,
		applicationshell_elementRoot_binding_1,
		applicationshell_elementContent_binding_1
	];
}

class DialogShell extends SvelteComponent {
	constructor(options) {
		super();

		init(this, options, instance, create_fragment, safe_not_equal, {
			elementContent: 0,
			elementRoot: 1,
			data: 3,
			dialogComponent: 2
		});
	}

	get elementContent() {
		return this.$$.ctx[0];
	}

	set elementContent(elementContent) {
		this.$$set({ elementContent });
		flush();
	}

	get elementRoot() {
		return this.$$.ctx[1];
	}

	set elementRoot(elementRoot) {
		this.$$set({ elementRoot });
		flush();
	}

	get data() {
		return this.$$.ctx[3];
	}

	set data(data) {
		this.$$set({ data });
		flush();
	}

	get dialogComponent() {
		return this.$$.ctx[2];
	}

	set dialogComponent(dialogComponent) {
		this.$$set({ dialogComponent });
		flush();
	}
}

export { ApplicationShell, DialogContent, DialogShell, EmptyApplicationShell, TJSApplicationHeader, TJSApplicationShell, TJSComponentShell, TJSContainer, TJSGlassPane, TJSHeaderButton };
